<!DOCTYPE html>
<html lang="en-us">

  <head>
  <link href="http://gmpg.org/xfn/11" rel="profile">
  <meta http-equiv="X-UA-Compatible" content="IE=edge">
  <meta http-equiv="content-type" content="text/html; charset=utf-8">

  <!-- Enable responsiveness on mobile devices-->
  <meta name="viewport" content="width=device-width, initial-scale=1.0, maximum-scale=1">

  <title>
    
      Lab 5 - Gene networks and Gene Ontology &middot; BIT150
    
  </title>

  <!-- CSS -->
  <link rel="stylesheet" href="/BIT150_website/public/css/poole.css">
  <link rel="stylesheet" href="/BIT150_website/public/css/syntax.css">
  <link rel="stylesheet" href="/BIT150_website/public/css/lanyon.css">
  <link rel="stylesheet" href="https://fonts.googleapis.com/css?family=PT+Serif:400,400italic,700%7CPT+Sans:400">

  <!-- Icons -->
  <link rel="logo" sizes="144x144" href="/BIT150_website/public/logo.png">
  <link rel="shortcut icon" href="/BIT150_website/public/favicon.ico">

  <!-- RSS -->
  <link rel="alternate" type="application/rss+xml" title="RSS" href="/atom.xml">
</head>


  <body class="theme-base-0g">

    <!-- Target for toggling the sidebar `.sidebar-checkbox` is for regular
     styles, `#sidebar-checkbox` for behavior. -->
<input type="checkbox" class="sidebar-checkbox" id="sidebar-checkbox" checked>

<!-- Toggleable sidebar -->
<div class="sidebar" id="sidebar">
  <div class="sidebar-item">
    <h1>
      <a href="/BIT150_website/">
        BIT150
      </a>
    </h1>
    <p class="lead">The UC Davis - CAES course for the development of fundamental skills in Bioinformatics</p>
  </div>

  <nav class="sidebar-nav">
    <a class="sidebar-nav-item" href="/BIT150_website/">Home</a>

    

    
    
      
        
      
    
      
        
      
    
      
        
          <a class="sidebar-nav-item" href="/BIT150_website/about/">About</a>
        
      
    
      
    
      
    
      
        
          <a class="sidebar-nav-item" href="/BIT150_website/course_syllabus/">BIT 150 Syllabus - Fall 2017</a>
        
      
    
      
        
          <a class="sidebar-nav-item" href="/BIT150_website/discussions/">Discussions</a>
        
      
    
      
    
      
        
          <a class="sidebar-nav-item" href="/BIT150_website/labs/">Labs</a>
        
      
    
      
        
          <a class="sidebar-nav-item" href="/BIT150_website/projects/">Projects</a>
        
      
    
      
        
          <a class="sidebar-nav-item" href="/BIT150_website/schedule/">Schedule</a>
        
      
    
      
        
          <a class="sidebar-nav-item" href="/BIT150_website/supplement/">Supplemental Material</a>
        
      
    

<!--
    <a class="sidebar-nav-item" href="/archive/v1.0.0.zip">Download</a>
    <a class="sidebar-nav-item" href="">GitHub project</a>
-->
    <span class="sidebar-nav-item">Currently v1.0.0</span>
  </nav>

  <div class="sidebar-item">
    <p>
      &copy; 2017, DE Runcie & SM Higdon. All rights reserved.
    </p>
  </div>
</div>


    <!-- Wrap is the content to shift when toggling the sidebar. We wrap the
         content to avoid any CSS collisions with our real content. -->
    <div class="wrap">
      <div class="masthead">
        <div class="container">
          <h3 class="masthead-title">
            <a href="/BIT150_website/" title="Home">BIT150</a>
            <small>Applied Bioinformatics</small>
          </h3>
        </div>
      </div>

      <div class="container content">
        <div class="post">
  <h1 class="post-title">Lab 5 - Gene networks and Gene Ontology</h1>
  <span class="post-date">26 Oct 2017</span>
  <div class="lab">
  <h1 id="goals">Goals</h1>

<ol>
  <li>Use Cytoscape to visualize a network</li>
  <li>Overlay data onto network topology</li>
  <li>Analyze a network:
    <ul>
      <li>topology</li>
      <li>GO enrichment</li>
    </ul>
  </li>
</ol>

<h1 id="to-learn-cytoscape-start-with-the-following-two-tutorials">To learn Cytoscape, start with the following two tutorials:</h1>

<h2 id="tutorial-1-intro-to-cytoscape">Tutorial 1. Intro to Cytoscape</h2>

<p>This tutorial will teach you to import a network into Cytoscape, adjust how it is rendered, and to overlay experimental data (in this case gene expression data) onto the nodes of the network.</p>

<h3 id="notes">Notes:</h3>
<p>One modification - instead of loading the file <code class="highlighter-rouge">galFiltered.sif</code>, load the file <code class="highlighter-rouge">galFiltered.csv</code> from the <code class="highlighter-rouge">Z:</code> drive. The file <code class="highlighter-rouge">galExpData.csv</code> is also on the <code class="highlighter-rouge">Z:</code> drive. Both files are also available on Canvas under <code class="highlighter-rouge">Lab_5</code>.</p>

<p><a href="http://manual.cytoscape.org/en/stable/Basic_Expression_Analysis_Tutorial.html">Basic Expression Analyis Tutorial</a></p>

<h2 id="tutorial-2-gene-ontology-enrichment-in-cytoscape">Tutorial 2. Gene Ontology enrichment in Cytoscape</h2>

<p>This tutorial will teach you to analyze a gene network for enrichments of gene ontology terms. Enrichments indicate that the set of genes in your network are non-random with respect to gene function. These “enriched” functions are good guesses for the overal importance of the network or pathway.</p>

<h3 id="notes-1">Notes:</h3>

<p>First, install the BiNGO App in Cytoscape. <code class="highlighter-rouge">Apps -&gt; App Manager</code>. Search for <code class="highlighter-rouge">BiNGO</code>. Click <code class="highlighter-rouge">Install</code>.</p>

<p>Start the tutorial with a new version of the galFiltered network. Re-import it as you did in the first tutorial into a new network collection.</p>

<p>Go only through Step 3. You can continue Steps 4+ later if you’d like.</p>

<p>When you do this for your own gene, be sure to change the organism if needed.</p>

<p><a href="https://www.psb.ugent.be/cbd/papers/BiNGO/Tutorial.html">GO Tutorial</a></p>

<h3 id="gene-identifiers">Gene Identifiers</h3>
<p>The BiNGO program tries to map the node IDs to GO terms using the following identifiers:
<img src="https://www.psb.ugent.be/cbd/papers/BiNGO/User_Guide_files/BiNGO_identifiers_per_organism.png" alt="" /></p>

<h3 id="interpretation-of-bingo-graph-and-results">Interpretation of BiNGO graph and results</h3>

<p>BiNGO is aimed at providing the user with a good idea of what functional themes are present in your gene set. The p-values give a good indication about the prominence of a given functional category. However, no biologist would dream of drawing conclusions solely based on p-values, and rightly so ! The p-values returned by BiNGO can give the user additional clues, which should be interpreted in the light of other evidence.</p>

<p>The BiNGO graph visualizes the GO categories that were found significantly over-represented in the context of the GO hierarchy. The size (area) of the nodes is proportional to the number of genes in the test set which are annotated to that node. The color of the node represents the (corrected) p-value. White nodes are not significantly over-represented, the other ones are, with a color scale ranging from yellow (p-value = significance level, e.g. 0.01) to dark orange (p-value = 5 orders of magnitude smaller than significance level, e.g. 10-5 * 0.01). The color saturates at dark orange for p-values which are more than 5 orders of magnitude smaller than the chosen significance level.</p>

<p>Due to the interdependency of functional categories in the GO hierarchy, it is very likely that not one category, but a whole branch of the GO hierarchy lights up as being significantly over-represented. In such cases, interpretation can be more difficult. The darkest orange nodes which are furthest down the hierarchy are probably the ones that you’re looking for. Suppose for example that a branch of metabolism categories lights up (see figure).</p>

<p><a href="https://www.psb.ugent.be/cbd/papers/BiNGO/User_Guide.html">More Backround Reading Here</a></p>

<h2 id="notes-2">Notes:</h2>

<ul>
  <li>To color the genes with a particular GO category, use the BiNGO output table to <strong>Select nodes</strong>. Then, in the gene network view, select the Style table, and click the <strong>third box</strong> in the style row you want to use (ex. <strong>Fill Color</strong>), and select a style. This box allows you to <strong>Bypass</strong> the normal attribute scheme just for the selected nodes.</li>
  <li>To get a table of the BiNGO results, in the BiNGO window, click the box <strong>Check Box for saving Data</strong>. This saves a <code class="highlighter-rouge">.bgo</code> file. This file can be opened in a text editor. You can copy the table from this file and paste it into Excel. From excel, you can then subset the columns you want.</li>
  <li>To use a table in Markdown, use this website: <a href="http://www.tablesgenerator.com/markdown_tables">Tables Generator</a>. It lets you paste in table data copied from Excel, and formats the table for Markdown.</li>
</ul>

<h2 id="other-networks">Other networks</h2>

<p>There are other ways to import networks into Cytoscape.</p>

<ol>
  <li>Download a edge file from a PPIN database like String or PSICQUIC.</li>
  <li><code class="highlighter-rouge">File -&gt; Import -&gt; Public Databases...</code>.
    <ul>
      <li>`Universal Interaction Database Client. This is PSICQuIC</li>
      <li><code class="highlighter-rouge">WikiPathways</code> a collection of community annotated networks and pathways.</li>
    </ul>
  </li>
  <li>Get the KEGGParser app from the app store. <code class="highlighter-rouge">Apps -&gt; KEGGParser -&gt; Load KGML -&gt; KEGG web load task</code>. You need to know the KEGG pathway ID first. Then you enter it here and it will download the network.</li>
</ol>

<p>You can use any of these networks for any of the tasks we described here.</p>

<h1 id="assignment">Assignment</h1>

<ol>
  <li>Create a network model for your gene for Project 2
    <ul>
      <li>Options: String netowrk, PSICQUIC network (can access from inside Cytoscape), WikiPathways network (can access from inside Cytoscape), or KEGG pathway (requires KEGGParser app inside Cytoscape)</li>
      <li>Import Network into Cytoscape</li>
      <li>Create a network view that shows the features of the network	- Identify over-represented gene functions within the network</li>
    </ul>
  </li>
  <li>Create a Markdown document including:
    <ul>
      <li>an image of the network</li>
      <li>A table of enriched GO terms</li>
      <li>A description of what you can learn from the network</li>
    </ul>
  </li>
</ol>

</div>

</div>

<div class="related">
  <h2>Related Posts</h2>
  <ul class="related-posts">
    
      <li>
        <h3>
          <a href="/BIT150_website/2017/11/30/lab-09/">
            Lab 9 - Genome Annotation
            <small>30 Nov 2017</small>
          </a>
        </h3>
      </li>
    
      <li>
        <h3>
          <a href="/BIT150_website/2017/11/16/lab-08/">
            Lab 8 - Genome Assembly
            <small>16 Nov 2017</small>
          </a>
        </h3>
      </li>
    
      <li>
        <h3>
          <a href="/BIT150_website/2017/11/09/lab-07/">
            Lab 7 - Genome-wide association study
            <small>09 Nov 2017</small>
          </a>
        </h3>
      </li>
    
  </ul>
</div>

      </div>
    </div>

    <label for="sidebar-checkbox" class="sidebar-toggle"></label>

    <script>
      (function(document) {
        var toggle = document.querySelector('.sidebar-toggle');
        var sidebar = document.querySelector('#sidebar');
        var checkbox = document.querySelector('#sidebar-checkbox');

        document.addEventListener('click', function(e) {
          var target = e.target;

          if(!checkbox.checked ||
             sidebar.contains(target) ||
             (target === checkbox || target === toggle)) return;

          checkbox.checked = false;
        }, false);
      })(document);
    </script>
  </body>
</html>
