---
layout: lab
title: Lab 8 - Genome Assembly
hidden: true
---

# Overview

* [Getting Setup](#getting-setup)
* [Illumina Paired End Reads](#illumina-paired-end-reads)
* [FastQ File Format](#fastq-file-format)
* [Genome Assembly](#genome-assembly)
* [Putting it all Together](#putting-it-all-together)
* [Lab 8 Assignment](#lab-8-assignment)

# Laboratory Objectives

* Familiarize concepts and principles related to genome sequencing
* Interpret genome assembly quality using standardized metrics
* Develop shell scripts that combine multiple programs for data analysis
* As part of the assignment, responses to all **Thought Questions** will be required

## Getting Setup

Today, you will analyze and process raw DNA sequence files from an *Escherichia coli* genome at the command line using programs on the **farm cluster**. Recall that in order to run computationally intensive commands on the farm cluster, we must first request the necessary compute space **outside of the login node** using the `srun` command.

1. Login to farm at the command line: `ssh user-id@farm.cse.ucdavis.edu` and enter your password.
2. Once you are logged in, request the resources for an interactive job to use in lab today:

  ```srun -p bit150 -t 2:00:00 --mem=8000 -n 2 --pty bash -l```

### Linking sequence data into a working directory

Once we have our interactive jobs running, the next step is to create a directory to work in for today's lab. We will put all files and subfolders associated with this lab into the directory `~/lab_08`:

```mkdir -p ~/Lab_8```

Next, let's create a `symbolic link` to each of the data files that we will be working with today from the `/groups/BIT150/Lab_8` folder. These two files contain the raw nucleotide reads for the Escherichia coli O104:H4 str. TY-2482 genome. To do this, we need to move into the `Lab_8` directory and then type the command to create symbolic links.

```
cd Lab_8
ln -s /group/BIT150/Lab_8/*.fastq.gz .
```

To verify that the symbolic links were created, type the following command:

`ls -l`

You should see output similar to the following:

```
total 0
lrwxrwxrwx 1 smhigdon smhigdon 40 Nov 15 22:11 SRR292770_1.fastq.gz -> /group/BIT150/Lab_8/SRR292770_1.fastq.gz
lrwxrwxrwx 1 smhigdon smhigdon 40 Nov 15 22:11 SRR292770_2.fastq.gz -> /group/BIT150/Lab_8/SRR292770_2.fastq.gz
```

These are FASTQ files - let's explore the meaning of the data contained within them...

## Illumina Paired End Reads

Illumina reads come in two styles - single read **(SR)** and paired end reads **(PE)**. **SR** read data of sequencing libraries is generally less expensive to generate, while **PE** read data is more costly to generate but contains additional information that is advantageous for generating genome assemblies. Illumina sequences are generated using the **sequencing by synthesis method:**

![]({{site.baseurl}}/images/lab8-1.png)

Paired end reads provide more information than single end reads because they provide sequence information pertaining to the reverse compliment of each sequence insert. This will often allow for the creation of a genome assembly with higher contiguity because the paired end reads can be used to resolve uncertainty in genomic regions of nucleotide residue homogeneity **(homopolymer repeats)**.

![]({{site.baseurl}}/images/lab8-2.png)


## FASTQ File format

[FASTQ Format](http://en.wikipedia.org/wiki/FASTQ_format){:target="_blank"} is the standard file format for Illumina sequence data. This file format is what you should currently expect to receive when downloading sequence data either from an online repository or from the file server of a DNA sequencing facility. However, fastq files are not the first form of the data that is generated from Illumina sequencers:

![]({{site.baseurl}}/images/lab8-3.png)

The fastq files contain data associated with each DNA insert read sequence generated from each PCR-amplified cluster on an Illumina flowcell. Each short read sequence record within the fastq files contain 4 lines:

![]({{site.baseurl}}/images/lab8-4.png)

Now that we have an idea of what a fastq file should look like, lets take a look at our files to make sure they meet the criteria for the fastq format:

    gunzip -c SRR292770_1.fastq.gz | less

**(use the spacebar to scroll down, and type 'q' to exit 'less')**

    gunzip -c SRR292770_2.fastq.gz | less

Notice that the two files have a common identifier that designates them as coming from the same sample. Now that we have confirmed the content of our files, let's move on to assembling the reads into a genome.

#### Thought Questions

>  * Why are there `_1` and `_2` in the file names?
>  * Why is each of the four lines from a fastq file read important?

## Genome Assembly

There are two main strategies for genome assembly using DNA sequencing reads, __*de novo* assembly__ and __reference based mapping__. Among the algorithm types used to carry out __*de novo* assembly__, the two most common strategies are __*De Bruijn Graph Assembly*__ and __*Overlap Layout Consensus*__. For more information on the differences between these two, here are links to some papers if you are interested:

* [Li et al., 2012](https://www.ncbi.nlm.nih.gov/pubmed/22184334){:target="_blank"}
* [Zerbino et al., 2008](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC2336801/){:target="_blank"}
* [Miller et al., 2010](http://www.sciencedirect.com/science/article/pii/S0888754310000492){:target="_blank"}

__*De Bruijn Graph Assembly*__ involves breaking all of the sequence reads down into smaller sequence fragments **(substrings)** of a specified length **(k-size)**, which are commonly referred to as **k-mers**.

![]({{site.baseurl}}/images/lab8-5.png)

The k-mers are typically ordered within a [hash table](https://en.wikipedia.org/wiki/Hash_table){:target="_blank"}, and a directional network graph is then constructed. Here is an example using a k-size of 3 that demonstrates the concept:

 ![]({{site.baseurl}}/images/lab8-6.png)

### MEGAhit

To assemble the *E. coli* genome reads we have been working with, we will use **MEGAhit**, which is an ultra-fast short read assembly program that takes a *De Bruijn Graph* approach. MEGAhit was designed for metagenome assembly, but it can also be used to assemble the shotgun whole genome sequence data from the DNA libraries of mono-isolate microbial genomes as well.

* The paper can be found [here](https://www.ncbi.nlm.nih.gov/pubmed/25609793){:target="_blank"}
* [MEGAhit source code](https://github.com/voutcn/megahit){:target="_blank"}

### k-mer size assembly activity

One question you may have thought of is, **"which k-mer size should be used for the assembly"** or, **"is there an optimal k-mer size for assembly?"** We will now do an in-class exercise together to investigate these questions.

Let's work together to answer these questions by assembling the read data using all possible k-mer sizes for the data that MEGAhit will allow. MEGAhit only allows for k-mer sizes of odd numbers.

**First**, click the following link to get to a live googlesheet for this exercise: [https://docs.google.com/spreadsheets/d/1Oxw2HLbRqEqduValTQ__I7atl23L6gMHZgwiwUj8UMU/edit#gid=0](https://docs.google.com/spreadsheets/d/1Oxw2HLbRqEqduValTQ__I7atl23L6gMHZgwiwUj8UMU/edit#gid=0){:target="_blank"}

Now, choose a k-mer size and fill in your name. **Our goal is to cover all k-mer sizes in this exercise, so if you see an empty slot, grab it.** Because the sequence read data we are working was generated using only 50 amplification cycles, we have really short reads and will only be able to cover 17 k-mer sizes during this exercise. If all of the k-mers have already been chosen, just choose the one you like the most, add your name in the column and continue with the exercise.

Once you have selected a k-mer size for your assembly, in the same directory containing the symbolic links to the fastq files we generated (~/lab_8), type the following command:

```
module load megahit
```

Recall that loading this module will connect your **PATH** environment variable to the directory with code for the MEGAhit program.

Now, type or copy the following command, and replace the *##* characters with your chosen k-mer size.

```
megahit --k-min ## --k-max ## --kmin-1pass -t 2 -1 SRR292770_1.fastq.gz -2 SRR292770_2.fastq.gz -o megahit_Ecoli_k##
```

After you enter this command, you will see lines of code begin to appear at the prompt.

> While the program is running, let's discuss some aspects of the assembly output MEGAhit will provide.

* Number of assembled **contigs** (contiguous sequences of DNA)
* Total assembly length (in bp)
* The N50 value
	* Note that one of the most common assembly statistics is the **N50** value.
	* The __N50__ value of an assembly is defined as the length of the smallest contig within a subset of the largest assembled contigs whose sum represents 50 percent of the total number of bases assembled.

When your assembly is complete, copy and paste the assembly statistics reported in the terminal by MEGAhit into the **googlesheet** for your chosen k-mer size.

You can also explore the assembly file generated with MEGAhit by typing the following command:

```
less -S megahit_Ecoli_k##/final.contigs.fa
```

** Again, replace the `##` characters with your chosen k-mer size.

#### Thought Questions

> * Why and how do different k-mer sizes affect the assembly results?
> * When would a short k-mer size be useful compared to the use of a long k-mer size?
> * How can the concept of k-mer size be related to `seed` size in BLAST?

MEGAhit provided us with a limited number of assembly statistics, and a question that may come to mind is...

> Are there other ways to assess the quality of my genome assembly?

The answer is yes, and to generate additional genome assembly metrics, we will use the program below.

## Assembly Quality

### QUAST: QUality ASsessment Tool for Genome Assemblies
* Click [here](http://quast.sourceforge.net/){:target="_blank"} for more information on QUAST.
* The paper can be found [here](https://academic.oup.com/bioinformatics/article/29/8/1072/228832/QUAST-quality-assessment-tool-for-genome){:target="_blank"}.
* [QUAST source code](https://github.com/ablab/quast){:target="_blank"}

To evaluate the quality of your MEGAhit assembly in more detail, descend into the directory containing the megahit output:

**substitute the `##` for the numbers you used before in the megahit command.**

```
cd megahit_Ecoli_k##
```

Next, load the `bio` module:

```
module load bio
```

> Recall  that we used the `bio` module when we were using BLAST at the command line in Lab 3. Loading this customized farm module will actually add many bioinformatics programs to your __$PATH__ so that you can easily use them in your interactive shell. To check out some of these programs, try typing the following commands after you have loaded the `bio` module:
* To view python modules: `pip list`
* To view [bioconda](https://bioconda.github.io/){:target="_blank"} modules: `conda list`

Now, execute the following command to run QUAST on your *E. coli* assembly:

```
quast.py final.contigs.fa -o quast_output
```

Running `QUAST` on the input file containing the *E. coli* TY-2482 assembly contigs from MEGAhit actually generated many files nested within the new directory, `quast_output`. Let's have a look at the report in plain text file:

```
less -S quast_output/report.txt
```

You should see something very similar to the following:

```
All statistics are based on contigs of size >= 500 bp, unless otherwise noted (e.g., "# contigs (>= 0 bp)" and "Tota

Assembly                    final.contigs
# contigs (>= 0 bp)         719
# contigs (>= 1000 bp)      369
# contigs (>= 5000 bp)      211
# contigs (>= 10000 bp)     151
# contigs (>= 25000 bp)     60
# contigs (>= 50000 bp)     19
Total length (>= 0 bp)      5210009
Total length (>= 1000 bp)   5065636
Total length (>= 5000 bp)   4702255
Total length (>= 10000 bp)  4259183
Total length (>= 25000 bp)  2792051
Total length (>= 50000 bp)  1285159
# contigs                   463
Largest contig              143131
Total length                5130667
GC (%)                      50.49
N50                         30790
N75                         14105
L50                         52
L75                         117
# N's per 100 kbp           0.00

```

> Note that another assembly statistic that is commonly referenced is the **L50** value, which is defined as the minimum number of contigs whose length sum is equal to the N50 value.**

#### Thought Questions

> * Why are there so many contigs?
> * How informative is the N50 value if you are comparing two genome assemblies that are different in total length?
> * What would be needed to produce an assembly with a total number of contigs under, say 10...or perhaps even a **single contig scaffold**?

## Putting it all together

Running through this entire process step by step is great for learning how everything works, but what if you need to assemble a large genome that will take a longer amount of time than, say five to ten minutes? A more efficient way to conduct this entire computational analysis is to write a shell script that can be submitted to SLURM with the `sbatch` command. Let's take a look at the example script below:

### sbatch script

    #!/bin/bash -l

    #SBATCH -J MEGA_B150
    #SBATCH -N 1
    #SBATCH -n 8
    #SBATCH --mem=16000
    #SBATCH -t 24:00:00

    # Name: megahit_BIT150.sh
    # Created by: Shawn Higdon
    # creation date: April 25, 2017
    # This is a pipeline for the assembly of Illumina PEx50 reads for individual microbial genomes.

    ## Load the required software module

    module load bio

    OUTPUT_FOLDER=lab_8
    mkdir -p $OUTPUT_FOLDER

    # Step One: Assemble the preprocessed read files for each sample using MEGAHIT

    MEGAHIT_FOLDER=$OUTPUT_FOLDER/megahit # make a megahit output folder

    megahit -1 $OUTPUT_FOLDER/*1.fq.gz -2 $OUTPUT_FOLDER/*2.fq.gz -o $MEGAHIT_FOLDER -t 8

    # Step Two: Use Quast to determine the quality of the MEGAHIT assembly

    QUAST_FOLDER=$MEGAHIT_FOLDER/quast

    quast.py $MEGAHIT_FOLDER/final.contigs.fa -o $QUAST_FOLDER

> Note the addition of the `sbatch` specific lines at the top of the script. Creating a directory called slurm-log at the specified path, and referencing it in the script with the `-e` and `-o` options tells `slurm` to generate error and output files, which aid in debugging your scripts. Adding the `--mail` options also activates the `slurm` feature of mailing the user when the job is complete.
>
> Having built in all of these `sbatch` commands, submitting this job is as simple as typing:
>
> `sbatch megahit_BIT150.sh`

## Lab 8 Assignment

For the assignment, create a markdown document that addresses the following and submit the document to canvas in **html** format:

### Part 1

Provide a response to each of the __*Thought Questions*__ that are presented at the end of each major section in this lab.

### Part 2

There are many Bioinformatics programs that have been written to carry out a __*Dr Bruijn Graph*__ based de novo assembly using paired-end Illumina reads. Here is a list of some assemblers that are already installed on farm:

> 1. [SPAdes](http://online.liebertpub.com/doi/abs/10.1089/cmb.2012.0021){:target="_blank"}
> 2. [Velvet](http://genome.cshlp.org/content/18/5/821.short){:target="_blank"}
> 3. [IDBA](https://link.springer.com/chapter/10.1007%2F978-3-642-12683-3_28){:target="_blank"}
> 4. [Abyss](http://genome.cshlp.org/content/19/6/1117.short){:target="_blank"}

Choose one of these programs and learn how it works by reading the documentation. Your task is to write an sbatch script that will assemble the same set of paired end reads you used in lab today using an assembler of your choice. Then, use **QUAST** to assess the assembly metrics of your new assembly.

> **Compare and contrast the outcome of your new assembly to the results of an assembly that was generated by running MEGAhit on the *E. coli* strain TY-2482 paired end reads with the default parameters. This data can be found on farm at the following path: `/group/BIT150/Lab_8/megahit_results`. Speculate as to why differences may or may not have occurred for each of the relevant assembly statistics.**

__*Be sure to include the code for your script in the markdown document.*__

To view the documentation for the assembly programs at the command line, type the following commands:

* Spades: `spades.py --h` 
* Velvet: `velveth`
* IDBA: `idba -h`
* Abyss: `man abyss`

__*Note:*__ Spades, Velvet and Abyss can all be accessed using, `module load bio`, which access to IDBA requires, `module load idba`.

Performing a google search is also a good place to start looking for usage examples.

----

> [Project 3](project-3.md){:target="_blank"}
