---
layout: lab
title: Lab 9 - Genome Annotation
hidden: true
---

# Overview

* [Getting Setup](#getting-setup)
* [Gene Prediction](#gene-prediction)
  * [PROKKA](#prokka)
* [Pan-genome analysis](#pan-genome-analysis-with-roary)
  * [Analysis in R](#visualizing-in-r)
* [Lab 9 Assignment](#lab-9-assignment)

# Laboratory Objectives

* Learn concepts and principles related to genome annotation
* Explore strategies for downstream analysis of genome annotations
* Develop shell scripts to automate the annotation of multiple assemblies
* As part of the assignment, responses to all **Thought Questions** will be required

# Getting Setup

Last week, we assembled the genome of *E. coli* O104:H4 strain TY-2482. This bacterium was identified as a virulent human pathogen that lead to a food-bourne disease outbreak in Europe. Today we will continue working with this organism by annotating its genome at the command line using programs on the **farm cluster**.

To begin:

1. Login to farm at the command line: `ssh user-id@farm.cse.ucdavis.edu` and enter your password.
2. Once you are logged in, request the following resources for an interactive job to use in lab today:

  ```
  srun -p bit150 -t 4:00:00 --mem=8000 -n 1 -c 3 --pty bash -l
  ```

# Gene Prediction

There are two major strategies for predicting genes within genome assemblies: reference based sequence alignment and *Ab initio* (from the beginning) - we will be exploring the latter. While there are pros and cons to both strategies, the *Ab initio* strategy is useful when a reference genome does not exist for the organism that has been sequenced. By this point, you are aware that genome features exist in several forms, including:

* Protein Coding sequences
* Transfer RNAs (tRNAs)
* Ribosomal RNAs (rRNAs)
* Non-coding RNAs (ncRNAs)

Each of these DNA feature types can be modeled based on the tetranucleotide patterns and frequencies associated their occurrence in nature. Consequently, there are several Bioinformatics programs that have been developed to *annotate* the stretches of genomic DNA and identify specific feature types. Today we will learn how to use the program **Prokka** to annotate our *E. coli* assembly. Prokka integrates a suite of genome annotation programs to comprehensively annotate a prokaryotic genome.

## Exercise

> * Describe potential benefits and limitations associated with using sequence alignment based techniques for genome annotation vs. *Ab initio* strategies.

## PROKKA

* [Prokka: rapid prokaryotic genome annotation](https://github.com/tseemann/prokka){:target="_blank"}
* [Prokka - iSCB Bioinformatics article](https://academic.oup.com/bioinformatics/article/30/14/2068/2390517){:target="_blank"}

Prokka is a bioinformatics program that is commonly used to annotate prokaryotic genomes using minimal computational resources. The major requirement of the input files is that they must be in *FASTA* format.

## Programs used to predict genome features

![]({{site.baseurl}}/images/prokka-1.png)

## A Simplified Schema of the PROKKA pipeline

![]({{site.baseurl}}/images/prokka-2.png)

## [Command Line Options](https://github.com/tseemann/prokka#command-line-options){:target="_blank"}

To view the command line options, we need to load the prokka module:

```
module load tbl2asn prokka
```

Now, type the following:

```
prokka
```

This should bring up a menu of the command line options that looks very similar to this:

![]({{site.baseurl}}/images/prokka-3.png){:target="_blank"}

## Annotating your *E. coli* genome assembly

Today you will annotate the assembled genome of *E. coli* O104:H4 strain TY-2482 that was generated last week.

First, create a folder in your home directory to work in for today's lab:

```
mkdir Lab_9
cd Lab_9
```

Similar to last week, __link-in__ the data from the data storage directory for the class:

```
ln -s /group/BIT150/Lab_8/megahit_results/final.contigs.fa ./ecoli_TY2482.contigs.fa
```

We re-name the file to `ecoli_TY2482.contigs.fa` because later we'll load assemblies of other bacteria, and we want to keep organized.

```
time prokka --outdir prokka_annotation --prefix E.coli_TY-2482 --cpus 3 ecoli_TY2482.contigs.fa
```
__*Note: Adding the time command before any command will print the compute time when the task is completed.*__

While this is running, let's discuss the output formats __(It should take about 5-10 minutes for PROKKA to finish)__.

## Output Formats

![]({{site.baseurl}}/images/prokka-4.png){:target="_blank"}

While many of the output formats serve the purpose of submitting data to **NCBI**, some useful formats for further analysis are the *FASTA amino acid* (**.faa**) and *GFF3 - General Feature Format 3* (**.gff**).

Let's explore the GFF3 output file together:

```
less -S prokka_annotation/E.coli_TY-2482.gff
```

Now, let's do something with the GFF3 file we just created!

## Exercise

> * __PROKKA__ annotates protein coding sequences using __BLAST__. What modifications to the `prokka` command would you have to make in order to increase the specificity of annotation assignments in your *E. coli* O104:H4 strain TY-2482 genome assembly?

## Pan Genome Analysis with Roary

* [Roary: The Pan Genome](http://sanger-pathogens.github.io/Roary/)
* [Roary - iSCB Bioinformatics Article](https://academic.oup.com/bioinformatics/article/31/22/3691/240757)

Now that we have annotated the genome of our virulent strain, and identified genes, we need a way to discover which genes might be the source of virulence. One way to do this is to compare the genes we found with genes in a closely related avirulent species.

**Roary** compares the annotation files of multiple prokaryotic genomes to identify a **Pan Genome**, which is the union of all genes identified among all the genomes. This is useful for identifying a set of **Core genes** present in all genomes, and a set of **Accessory Genes** which are present in some, but not all genomes.

We've included the **GFF3** file for the *avirulent reference* strain of *E. coli O104:H4*, which is closely related to TY-2482 in the `/group/BIT150/Lab_9` folder.

To begin the Pan Genome analysis, let's make a new sub-directory inside the `Lab_9` directory and move into it:

```
mkdir roary
cd roary
```

Now, __link-in__ the **GFF3** file for the *avirulent reference* strain of *E. coli O104:H4* from the `/group/BIT150/Lab_9` folder, and the **GFF3** file from your **prokka_annotation** folder to your present working directory (roary):

```
ln -s /group/BIT150/Lab_9/E.coli_O104-H4.gff .
ln -s ../prokka_annotation/E.coli_TY-2482.gff .
```

At this point, you should have **two GFF files** in your present working directory - check by typing `ls`.

To begin using the program, we need to load another module:

```
module load bio3
```

Next, let's bring up the roary options menu:

```
roary
```

As you can see, roary has many options for customizing your pan-genome analysis. Today we will keep things fairly straight forward. Run the following command to generate the core gene alignment and associated pan genome output files (**Note**:*This takes about 20 minutes* to finish):

```
time roary -i 90 -e --mafft -p 3 *.gff
```

> While this is running, let's explore the genome annotation of our reference *E. coli* **together** using the **Integrated Genome Browser**.


Once your Pan Genome analysis is completed, check out the statistics:

```
less summary_statistics.txt
```
**q to exit less window**

The main file we want to investigate for our comparative analysis is in the `gene_presence_absence.csv` file.

Have a look:

```
less gene_presence_absence.csv
```

As you can see, the data is easy to interpret here. Let's analyze the data in R and make a plot.

## Visualizing in R

We will use R Studio to analyze the data. **The objective is to create a simple Venn Diagram** that shows the number of unique genes in the virulent strain of *E. coli*, the number of unique genes in the avirulent strain of *E. coli*, and the number of shared genes between them (the Core genome between the two).

### Getting Started
First, Launch R Studio, then create a new R project (essentially an entire directory dedicated to a single R analysis project):

> **(File -> New Project -> New Directory -> Empty Project -> *Name your directory something like R_roary*)**.

Now, we need to get the file from the HPC server onto our local computers. Launch **File Zilla**, connect to farm as you have done in the past, and transfer the `gene_presence_absence.csv` file to the folder on your workstation's hard-drive that corresponds to the **R project directory** you just created. Once this file is transferred, it should be visible in the R Studio window's file browser:

![]({{site.baseurl}}/images/lab9-R-1.png){:target="_blank"}

Create a New Rscript (.R), and save the file with a descriptive name - this file is where you will save all of your work:

> **(File -> New File -> R Script)**

**Note**: *Recall that commands are run from the R script editor by holding the __control__ key and pressing __enter__ on a given line or highlighted text*

The R package we will use to make the Venn Diagram is called, [VennDiagram](https://rstudio-pubs-static.s3.amazonaws.com/13301_6641d73cfac741a59c0a851feb99e98b.html){:target="_blank"}. We will also use the **dplyr** package to filter our dataset.

Load the necessary R Package libraries:

```
library(dplyr)
install.packages('VennDiagram')
library(VennDiagram)
```

To begin, read in the *csv* file as a dataframe, treating the first line as a header and filling in all blank cells with an "NA":

```
roary_data <- read.csv("./gene_presence_absence.csv", header = T, na.strings=c("","NA"))
```

Recall that you may view the dataframe using the `View()` command or by clicking the table icon of the `R Environment` pane of the R Studio session.

### Filter the Data
The next step is to identify the number of Accessory Genome genes that correspond to either *E. coli* O104:H4 strain TY-2482, or the reference strain that is avirulent. To do this, we will use a series of filtering steps in order to produce three new dataframes that represent subsets of the entire roary pan genome dataset that we generated.


* __Subset the Core genome:__ Filter by genes that are found in both genotypes (`No..isolates` column == 2)

```
core <- roary_data %>% filter(No..isolates == 2)
```


* __Subset the Accessory genome:__ Filter by genes that are found in only 1 of the 2 genotypes (`No..isolates` column == 1)

```
accessory_genome <- roary_data %>% filter(No..isolates == 1)
```

* __Subset the TY-2482 Accessory genome:__ Filter by genes that are not `NA` in the `E.coli_TY.2842` column

```
TY2482_only <- accessory_genome %>% filter(!is.na(E.coli_TY.2842))
```

* __Subset the Accessory genome of the Reference genome:__ Filter by genes that are not `NA` in the `E.coli_ref ` column

```
reference_only <- accessory_genome %>% filter(!is.na(E.coli_ref))
```

### Making the Venn Diagram

Each row in the subsetted datasets represents an annotated gene. Thus, we can count the row numbers of each genome subset, store the values in R-type objects and call on these R environment objects to make our plot:

* __Count the number of rows in each Pan genome subset:__

```
core_size <- nrow(core)
TY_2482_accessory_size <- nrow(TY_2482_only)
ref_accessory_size <- nrow(reference_only)
```

* __Make the diagram:__

```
grid.newpage()
draw.pairwise.venn(
  area1 = TY_2482_accessory_size + core_size, # count # genes in TY2482
  area2 = ref_accessory_size + core_size, # count # genes in Ecoli_ref
  cross.area = core_size, # count number of genes in both
  category = c("Virulent", "Avirulent"),
  fill = c("blue", "red")
  )
```

Here we are visualizing the Pan-genome in a simplistic fashion, but this opens the door to further exploration of each isolate's contribution to the accessory genome of the comparison.

### Exercise

 > Find a gene or two that you think might be relevant to the virulence of the organism. Explain your reasoning.

## Lab 9 Assignment

For the assignment, create a markdown document that addresses the following and submit the document to canvas in **html** format:

### Part 1: Exercise Questions

Provide a response to each of the three __*Exercise Questions*__ that are presented at the end of major sections in this lab.

### Part 2: RAST Annotation

A very widely used web-based tool for microbial genome annotation is **RAST**. Your task is to figure out how to use this program (which is similar to Prokka), use it to re-annotate the `E.coli_TY.2842` assembly, and then use `roary` again to identify *Core* and genome-specific genes for the two bacteria. Present a Venn diagramm of the results of the RAST annotation.

The RAST program is available here: [RAST](http://rast.nmpdr.org/)

**Note:** You have to register for an account to use RAST. Please register today (it might take a day to get accepted!)

----

[Project 3](project-3.md)
