<!DOCTYPE html>
<html lang="en-us">

  <head>
  <link href="http://gmpg.org/xfn/11" rel="profile">
  <meta http-equiv="X-UA-Compatible" content="IE=edge">
  <meta http-equiv="content-type" content="text/html; charset=utf-8">

  <!-- Enable responsiveness on mobile devices-->
  <meta name="viewport" content="width=device-width, initial-scale=1.0, maximum-scale=1">

  <title>
    
      Lab 3 - HPC and Command Line BLAST &middot; BIT150
    
  </title>

  <!-- CSS -->
  <link rel="stylesheet" href="/BIT150_website/public/css/poole.css">
  <link rel="stylesheet" href="/BIT150_website/public/css/syntax.css">
  <link rel="stylesheet" href="/BIT150_website/public/css/lanyon.css">
  <link rel="stylesheet" href="https://fonts.googleapis.com/css?family=PT+Serif:400,400italic,700%7CPT+Sans:400">

  <!-- Icons -->
  <link rel="logo" sizes="144x144" href="/BIT150_website/public/logo.png">
  <link rel="shortcut icon" href="/BIT150_website/public/favicon.ico">

  <!-- RSS -->
  <link rel="alternate" type="application/rss+xml" title="RSS" href="/atom.xml">
</head>


  <body class="theme-base-0g">

    <!-- Target for toggling the sidebar `.sidebar-checkbox` is for regular
     styles, `#sidebar-checkbox` for behavior. -->
<input type="checkbox" class="sidebar-checkbox" id="sidebar-checkbox" checked>

<!-- Toggleable sidebar -->
<div class="sidebar" id="sidebar">
  <div class="sidebar-item">
    <h1>
      <a href="/BIT150_website/">
        BIT150
      </a>
    </h1>
    <p class="lead">The UC Davis - CAES course for the development of fundamental skills in Bioinformatics</p>
  </div>

  <nav class="sidebar-nav">
    <a class="sidebar-nav-item" href="/BIT150_website/">Home</a>

    

    
    
      
        
      
    
      
        
      
    
      
        
          <a class="sidebar-nav-item" href="/BIT150_website/about/">About</a>
        
      
    
      
    
      
    
      
        
          <a class="sidebar-nav-item" href="/BIT150_website/course_syllabus/">BIT 150 Syllabus - Fall 2017</a>
        
      
    
      
        
          <a class="sidebar-nav-item" href="/BIT150_website/discussions/">Discussions</a>
        
      
    
      
    
      
        
          <a class="sidebar-nav-item" href="/BIT150_website/labs/">Labs</a>
        
      
    
      
        
          <a class="sidebar-nav-item" href="/BIT150_website/projects/">Projects</a>
        
      
    
      
        
          <a class="sidebar-nav-item" href="/BIT150_website/schedule/">Schedule</a>
        
      
    
      
        
          <a class="sidebar-nav-item" href="/BIT150_website/supplement/">Supplemental Material</a>
        
      
    

<!--
    <a class="sidebar-nav-item" href="/archive/v1.0.0.zip">Download</a>
    <a class="sidebar-nav-item" href="">GitHub project</a>
-->
    <span class="sidebar-nav-item">Currently v1.0.0</span>
  </nav>

  <div class="sidebar-item">
    <p>
      &copy; 2017, DE Runcie & SM Higdon. All rights reserved.
    </p>
  </div>
</div>


    <!-- Wrap is the content to shift when toggling the sidebar. We wrap the
         content to avoid any CSS collisions with our real content. -->
    <div class="wrap">
      <div class="masthead">
        <div class="container">
          <h3 class="masthead-title">
            <a href="/BIT150_website/" title="Home">BIT150</a>
            <small>Applied Bioinformatics</small>
          </h3>
        </div>
      </div>

      <div class="container content">
        <div class="post">
  <h1 class="post-title">Lab 3 - HPC and Command Line BLAST</h1>
  <span class="post-date">12 Oct 2017</span>
  <div class="lab">
  <h1 id="outline">Outline:</h1>
<ul>
  <li>review of for loops</li>
  <li>Intro to using a computer cluster</li>
  <li>Running BLAST from the command line</li>
  <li>Running more complex BLAST jobs</li>
</ul>

<h1 id="review-of-loops">Review of loops</h1>
<p>In the homework, I asked you to write a script to process all pairs of files through the <code class="highlighter-rouge">goodiff</code> program. This requires <strong>nested for loops</strong></p>

<p>Let’s start in the <code class="highlighter-rouge">north-pacific-gyre/2012-07-03</code> data files folder.</p>

<p>We can see all the files we want to process with:</p>

<div class="highlighter-rouge"><pre class="highlight"><code>ls *.txt
</code></pre>
</div>

<p><strong>Note: some of you may have other <code class="highlighter-rouge">.txt</code> files in this folder. If you do, modify the above command to only list the data files</strong></p>

<p>To process all pairs of files, we want to create two variables to hold the names of the two files to process: <code class="highlighter-rouge">file1</code> and <code class="highlighter-rouge">file2</code>. Then we want to use <strong>for loops</strong> to substitute each possible pair of files for these two variables.</p>

<p>First, lets start by processing the first file against all others. You could use a script like this:</p>

<div class="highlighter-rouge"><pre class="highlight"><code>file1=NENE01729A.txt
for file2 in NE*.txt
do
 echo Processing $file1 $file2
 programs/goodiff $file1 $file2 &gt;&gt; output/goodiff_results.txt
done
</code></pre>
</div>

<p>No, we want to run this whole script again and again, but each time with a different <code class="highlighter-rouge">file1</code>. We can do this by writing a loop for file1, where the <strong>body</strong> of the loop is this first loop:</p>

<div class="highlighter-rouge"><pre class="highlight"><code>for file1 in NE*.txt
do
  echo $file1
  for file2 in NE*.txt
  do
   echo Processing $file1 $file2
   programs/goodiff $file1 $file2 &gt;&gt; output/goodiff_results.txt
  done
done
</code></pre>
</div>

<p>That’s it!</p>

<h1 id="intro-to-high-performance-computing">Intro to High Performance Computing</h1>

<h3 id="key-ideas">Key ideas:</h3>

<ul>
  <li>The Farm cluster is a <strong>shared resource</strong>. It is important to follow some rules so that you do not interfere with other users.</li>
  <li>The Farm cluster has a head node and then a bunch of compute nodes
    <ul>
      <li>You log in from your computer to the head node, and then ask for time on a compute node.</li>
      <li>You have to move files from your computer to the cluster. Once there, the whole file system is shared among all nodes. You don’t need to keep copying files</li>
      <li>The cluster runs Linux (clone of Unix). You control it using Bash</li>
      <li>There are lots of programs installed, and different versions of many programs. Use <code class="highlighter-rouge">modules</code> to select which programs you want to use. It takes care of the rest.</li>
      <li>You can log in to Farm from anywhere as long as you have your username / password. This will be given to you (it’s not your Kerberos login).</li>
    </ul>
  </li>
</ul>

<h3 id="log-in">Log in:</h3>
<p>To log in, from a Unix terminal type:</p>

<div class="highlighter-rouge"><pre class="highlight"><code>ssh username@farm.cse.ucdavis.edu
</code></pre>
</div>

<p>Then enter your password. You will arrive at your personal <strong>home directory</strong>, which will be: <code class="highlighter-rouge">/home/username</code>. Only you can write / modify files in this directory.</p>

<h3 id="directories">Directories:</h3>
<p>There is a class directory called: <code class="highlighter-rouge">/group/BIT150</code>. You can access these files, but they are locked - you cannot change them. If you want to share files with the class, you can put them in: <code class="highlighter-rouge">/group/BIT150/Share</code>. Change to each of these directories (using <code class="highlighter-rouge">cd</code>), and view their contents (using <code class="highlighter-rouge">ls</code>).</p>

<h3 id="moving-files-on-and-off">Moving files on and off:</h3>
<p>Open the <code class="highlighter-rouge">FileZilla</code> program. This program sets up a secure connection to the Farm file server and lets you move files on and off of the cluster. It works a lot like <code class="highlighter-rouge">Windows File Explorer</code>.</p>

<p>To use:</p>

<ol>
  <li>Select new site, enter: Host: <code class="highlighter-rouge">farm.cse.ucdavis.edu</code>, Port: <code class="highlighter-rouge">2022</code>, Protocol: <code class="highlighter-rouge">SFTP</code>, Logon Type: “Normal”, User: <code class="highlighter-rouge">username</code>, Password: <code class="highlighter-rouge">password</code>, Click <code class="highlighter-rouge">Connect</code>.</li>
  <li>On the left is a file explorer for your computer. Navigate to the <code class="highlighter-rouge">Z:/Classes/BIT150/Lab_3</code>.</li>
  <li>On the right is a file explorer for Farm’s file system. Create a directory called <code class="highlighter-rouge">Lab_3</code>. You can also do this in the terminal (<code class="highlighter-rouge">mkdir Lab_3</code>).</li>
  <li>Find the file: <code class="highlighter-rouge">brca1.fasta</code>, and drag it onto Farm.</li>
  <li>In the terminal, enter the command: <code class="highlighter-rouge">echo Some Text &gt; new_file.txt</code>. In <code class="highlighter-rouge">FileZilla</code>, refresh the Farm file explorer. The file <code class="highlighter-rouge">new_file.txt</code> should appear. Double-click, or drag it to your computer.</li>
</ol>

<h2 id="running-blast">Running BLAST</h2>

<p>The BLAST programs are installed on Farm. They are fairly intensive programs, requiring a moderate amount of RAM, and sometimes a lot of time to run.</p>

<p><strong>We do not want to run BLAST on the head node!</strong> This could bog-down this computer and prevent others on the cluster from doing their work!</p>

<p>First, request an allocation on one of the compute nodes:</p>

<h3 id="requesting-an-interactive-job">Requesting an interactive job</h3>
<div class="highlighter-rouge"><pre class="highlight"><code>srun -p bit150 -t 20 --pty bash -l
</code></pre>
</div>
<ul>
  <li><code class="highlighter-rouge">srun</code> is a command that asks for an allocation</li>
  <li><code class="highlighter-rouge">-p bit150</code> asks to use the <code class="highlighter-rouge">bit150</code> partition. This means use the resources assigned to our class</li>
  <li><code class="highlighter-rouge">-t 20</code> asks for 20 minutes. After this time, your allocation will shut down</li>
  <li><code class="highlighter-rouge">--pty bash -l</code> asks for the allocation to run <code class="highlighter-rouge">Bash</code> just as if you opened a new terminal</li>
</ul>

<p>Hit <code class="highlighter-rouge">Enter</code>, and then wait a minute. Eventually you will see some messages saying you are logged in to a job.</p>

<p>To see info on your job, type:</p>
<div class="highlighter-rouge"><pre class="highlight"><code>squeue -u username
</code></pre>
</div>

<h3 id="the-blast-program">The blast program</h3>

<p>Now that we have access to a powerful computer, we can use BLAST to search a database for our sequence.</p>

<p>Our sequence is in the file <code class="highlighter-rouge">brca1.fasta</code>, the file we moved to Farm. View the file with:</p>

<div class="highlighter-rouge"><pre class="highlight"><code>head brca1.fasta
</code></pre>
</div>

<h4 id="question-is-this-a-nucleotide-of-protein-sequence">Question: Is this a nucleotide of protein sequence?</h4>

<h3 id="the-fasta-format">The FASTA format</h3>

<p>The FASTA	format is a common format for storing sequence files. Each sequence has a header (<strong>the metadata</strong>), and then the sequence. The header is the first line, and begins with the <code class="highlighter-rouge">&gt;</code> character. The rest of the line contains various annotations (a very compressed version of the GenBank format). Longer sequences are split over multiple lines. But the line breaks are irrelevant. Programs like BLAST know to keep reading the sequence until the end of the file.</p>

<p>Multiple sequences can be held in the same file. They are separated by header lines.</p>

<p>First, we have to load the module with the blast program. The bio module includes many bioinformatics programs. Type:</p>

<div class="highlighter-rouge"><pre class="highlight"><code>module load bio
</code></pre>
</div>

<p>Now, we can run blast. A run looks like this:</p>

<div class="highlighter-rouge"><pre class="highlight"><code>blastp -db /group/BIT150/blast_databases/refseq_human_protein -query brca1.fasta -out brca1_blast.txt
</code></pre>
</div>

<ul>
  <li><code class="highlighter-rouge">blastp</code> is the program for protein-protein blast.</li>
  <li><code class="highlighter-rouge">db /group/BIT150/blast_databases/refseq_human_protein</code> specifies the blast database that we want to use. We select the <code class="highlighter-rouge">refseq_human_protein </code> database in the <code class="highlighter-rouge">/group/BIT150/blast_databases/</code> directory</li>
  <li><code class="highlighter-rouge">query brca1.fasta</code> specifies the query sequence</li>
  <li><code class="highlighter-rouge">out brca1_blast.txt </code> specifes we want the output stored in the file called <code class="highlighter-rouge">brca1_blast.txt </code></li>
</ul>

<p>Look at the file. It’s long. A good viewer is <code class="highlighter-rouge">less</code>. Or, you can download it to your computer using <code class="highlighter-rouge">FileZilla</code>, and view it in <code class="highlighter-rouge">Notepad++</code>.</p>

<p>Let’s modify the command. We can change the output style into a concise table. Rather than re-typing, use the up-arrow to get find the command. <strong>Note:</strong> or, you could use <code class="highlighter-rouge">history</code>. Now, instead of moving the cursor with the arrow keys, copy it to a Notepad++ document, and edit it there.</p>

<p>For clarity, it’s often nice to break up long commands into several lines. You can do this by putting a <code class="highlighter-rouge">\</code> character at the end of each line.</p>

<div class="highlighter-rouge"><pre class="highlight"><code>blastp  \
   -db /group/BIT150/blast_databases/refseq_human_protein \
   -query brca1.fasta \
   -out brca1_blast_table.txt \
   -outfmt 7   
</code></pre>
</div>

<p>I added the option <code class="highlighter-rouge">-outfmt 6</code>. This changes the format of the output to a table. See <code class="highlighter-rouge">blastp -help</code> for more formatting options.</p>

<p>You can copy this whole set of code and paste into the terminal. Now, look at this output file using <code class="highlighter-rouge">less</code>.</p>

<p>When you’re done, exit the compute node by typing:</p>

<div class="highlighter-rouge"><pre class="highlight"><code>exit
</code></pre>
</div>

<p><strong>Always remember to exit when you’re done - this frees up resources for others!</strong></p>

<h4 id="notes">Notes:</h4>

<ul>
  <li>There are many more options for the <code class="highlighter-rouge">blastp</code> program. You can view them with <code class="highlighter-rouge">blastp -help</code>, or <a href="https://www.ncbi.nlm.nih.gov/books/NBK279675/">here</a></li>
  <li>There are many databases that you could blast against. You can also create your own starting with a file of fasta sequences using the <code class="highlighter-rouge">makeblastdb</code> command. <a href="https://www.ncbi.nlm.nih.gov/books/NBK279688/">Link</a></li>
  <li>You can extract sequences from a database with the <code class="highlighter-rouge">blastdbcmd</code> program <a href="https://www.ncbi.nlm.nih.gov/books/NBK279689/">Link</a></li>
</ul>

<h2 id="running-longer-jobs-with-sbatch">Running longer jobs with sbatch</h2>

<p>The <code class="highlighter-rouge">srun</code> command is useful for testing commands and shorter programs. But it has limitations for longer or more intensive jobs.</p>

<ul>
  <li>Sometimes it takes a long time to get an allocation. With <code class="highlighter-rouge">srun</code> you have to wait.</li>
  <li>With <code class="highlighter-rouge">srun</code>, you have to wait until your program is done and then remember to log out.</li>
</ul>

<p>The preferred way to run computation on a cluster like Farm is by submitting jobs that run <strong>non-interactively</strong> using the command <code class="highlighter-rouge">sbatch</code>.</p>

<p>Launching a <code class="highlighter-rouge">sbatch</code> job looks like this:</p>

<div class="highlighter-rouge"><pre class="highlight"><code>sbatch -p bit150 -t 30 my_script.sh
</code></pre>
</div>

<p>This will tell the cluster to find a time to allocate resources to my job for 30 minutes, and then run the script <code class="highlighter-rouge">my_script.sh</code></p>

<p>The key to using <code class="highlighter-rouge">sbatch</code> is writing a good script. The body of the script is the set of commands you would type in to Bash if you were there to control it. But you start the script with some special commands to tell the cluster how to run your job.</p>

<p>The simplest script looks like this:</p>

<div class="highlighter-rouge"><pre class="highlight"><code><span class="c">#!/bin/bash</span>

module load bio

blastp  <span class="se">\</span>
   -db /group/BIT150/blast_databases/refseq_human_protein <span class="se">\</span>
   -query brca1.fasta <span class="se">\</span>
   -out brca1_blast_table.txt <span class="se">\</span>
   -outfmt 7    
</code></pre>
</div>

<ul>
  <li>The first line tells the cluster to run this script with the <code class="highlighter-rouge">bash</code> shell.</li>
  <li>The second tells it to load the bio module</li>
  <li>Then we give the command for running blast.</li>
</ul>

<p>We can then modify this script to do longer, more intensive jobs:</p>

<p>The following script does a blast using a query file that includes 10 different sequences:</p>

<div class="highlighter-rouge"><pre class="highlight"><code><span class="c">#!/bin/bash</span>

module load bio

blastp  <span class="se">\</span>
   -db /group/BIT150/blast_databases/refseq_human_protein <span class="se">\</span>
   -query /group/BIT150/Lab_3/mouse_protein_set1.fasta <span class="se">\</span>
   -out mouse_protein_set1_blast_table.txt <span class="se">\</span>
   -outfmt 7    
</code></pre>
</div>

<h2 id="modular-scripts">Modular scripts</h2>
<p>The above script is fine, but it is a bit hard to read, and is also hard to modify. Say you wanted to blast a different gene against a different database? How would you change the script?</p>

<p>The way to write cleaner scripts is to use bash variables. We saw variables earlier when writing for loops. The concept here is the same. We can assign a <strong>value</strong> to a <strong>variable</strong> in the top of the script, and then use the variable later:</p>

<div class="highlighter-rouge"><pre class="highlight"><code><span class="c">#!/bin/bash</span>

module load bio

<span class="c"># declare database and query files</span>
<span class="nv">database</span><span class="o">=</span>/group/BIT150/blast_databases/refseq_human_protein
<span class="nv">query</span><span class="o">=</span>/group/BIT150/Lab_3/mouse_protein_set1.fasta
<span class="nv">output_file</span><span class="o">=</span>mouse_protein_set1_blast_table.txt

blastp  <span class="se">\</span>
   -db <span class="nv">$database</span> <span class="se">\</span>
   -query <span class="nv">$query</span> <span class="se">\</span>
   -out mouse_protein_set1_blast_table.txt <span class="se">\</span>
   -outfmt 7
</code></pre>
</div>

<p>Now the call to <code class="highlighter-rouge">blastp</code> is much clearer. It’s also often helpful to separate directories from file names, and base names from file extensions. A better script might look like this:</p>

<div class="highlighter-rouge"><pre class="highlight"><code><span class="c">#!/bin/bash</span>

module load bio

<span class="c"># declare database directory</span>
<span class="nv">database_dir</span><span class="o">=</span>/group/BIT150/blast_databases/

<span class="c"># declare input file directory</span>
<span class="nv">query_dir</span><span class="o">=</span>/group/BIT150/Lab_3

<span class="c"># declare output file directory</span>
<span class="nv">output_dir</span><span class="o">=</span>.

<span class="c"># database</span>
<span class="nv">database</span><span class="o">=</span>refseq_human_protein

<span class="c"># query</span>
<span class="nv">query</span><span class="o">=</span>mouse_protein_set1

blastp  <span class="se">\</span>
   -db <span class="nv">$database_dir</span>/<span class="nv">$database</span> <span class="se">\</span>
   -query <span class="nv">$query_dir</span>/<span class="k">${</span><span class="nv">query</span><span class="k">}</span>.fasta <span class="se">\</span>
   -out <span class="nv">$output_dir</span>/<span class="k">${</span><span class="nv">query</span><span class="k">}</span>_blast_table.txt <span class="se">\</span>
   -outfmt 7
</code></pre>
</div>

<h3 id="activities">Activities</h3>

<ol>
  <li>Modify the above script to blast the <code class="highlighter-rouge">mouse_protein_set1.fasta</code> file against three of these databases:</li>
</ol>

<div class="highlighter-rouge"><pre class="highlight"><code>refseq_human_protein
refseq_mouse_protein
refseq_zebrafish_protein
refseq_plant_protein
swissprot
</code></pre>
</div>

<ol>
  <li>Modifying the above script, write a <strong>for loop</strong> to blast each of the files in <code class="highlighter-rouge">/group/BIT150/Lab_3/individual_genes</code> against one of the protein databases, saving each in it’s own file in an output directory of your home directory?</li>
</ol>

<h2 id="notes-1">Notes</h2>

<p>When submitting <code class="highlighter-rouge">sbatch</code> jobs, two important commands are:</p>

<ul>
  <li><code class="highlighter-rouge">squeue -p bit150</code>, or <code class="highlighter-rouge">squeue -u bit150-01</code>. These show you if your job is running</li>
  <li><code class="highlighter-rouge">scancel -u bit150-01</code>. This cancels all your running jobs. You can also do: <code class="highlighter-rouge">scancel 15968444</code> to cancel a specific job numer</li>
</ul>

<h1 id="homework">Homework:</h1>
<ol>
  <li>Find the amino acid sequence of your gene</li>
  <li>Choose three of the blast databases on Farm. Here is a list:
    <blockquote>

      <table>
        <thead>
          <tr>
            <th>Database</th>
            <th>Type</th>
            <th>program</th>
          </tr>
        </thead>
        <tbody>
          <tr>
            <td><code class="highlighter-rouge">refseq_human_protein</code></td>
            <td>Protein</td>
            <td>blastp</td>
          </tr>
          <tr>
            <td><code class="highlighter-rouge">refseq_mouse_protein</code></td>
            <td>Protein</td>
            <td>blastp</td>
          </tr>
          <tr>
            <td><code class="highlighter-rouge">refseq_zebrafish_protein</code></td>
            <td>Protein</td>
            <td>blastp</td>
          </tr>
          <tr>
            <td><code class="highlighter-rouge">refseq_plant_protein</code></td>
            <td>Protein</td>
            <td>blastp</td>
          </tr>
          <tr>
            <td><code class="highlighter-rouge">swissprot</code></td>
            <td>Protein</td>
            <td>blastp</td>
          </tr>
          <tr>
            <td><code class="highlighter-rouge">Cdd_NCBI</code></td>
            <td>Domains</td>
            <td>rpsblast</td>
          </tr>
          <tr>
            <td><code class="highlighter-rouge">Pfam</code></td>
            <td>Domains</td>
            <td>rpsblast</td>
          </tr>
          <tr>
            <td><code class="highlighter-rouge">Smart</code></td>
            <td>Domains</td>
            <td>rpsblast</td>
          </tr>
        </tbody>
      </table>
    </blockquote>
  </li>
  <li>
    <p>Write a script to BLAST your sequence against these three databases. Note that the last three are databases of protein domains. These require the <code class="highlighter-rouge">rpsblast</code> program instead of <code class="highlighter-rouge">blastp</code></p>
  </li>
  <li>Now, go to NCBI’s blast server and run the same searches there. Compare the results.</li>
  <li>Write a Markdown document describing what you learned about your gene. Include your script from step 3.</li>
</ol>

</div>

</div>

<div class="related">
  <h2>Related Posts</h2>
  <ul class="related-posts">
    
      <li>
        <h3>
          <a href="/BIT150_website/2017/11/16/lab-08/">
            Lab 8 - Genome Assembly
            <small>16 Nov 2017</small>
          </a>
        </h3>
      </li>
    
      <li>
        <h3>
          <a href="/BIT150_website/2017/11/09/lab-07/">
            Lab 7 - Genome-wide association study
            <small>09 Nov 2017</small>
          </a>
        </h3>
      </li>
    
      <li>
        <h3>
          <a href="/BIT150_website/2017/11/02/lab-06/">
            Lab 6 - Genome Browsers
            <small>02 Nov 2017</small>
          </a>
        </h3>
      </li>
    
  </ul>
</div>

      </div>
    </div>

    <label for="sidebar-checkbox" class="sidebar-toggle"></label>

    <script>
      (function(document) {
        var toggle = document.querySelector('.sidebar-toggle');
        var sidebar = document.querySelector('#sidebar');
        var checkbox = document.querySelector('#sidebar-checkbox');

        document.addEventListener('click', function(e) {
          var target = e.target;

          if(!checkbox.checked ||
             sidebar.contains(target) ||
             (target === checkbox || target === toggle)) return;

          checkbox.checked = false;
        }, false);
      })(document);
    </script>
  </body>
</html>
