<!DOCTYPE html>
<html lang="en-us">

  <head>
  <link href="http://gmpg.org/xfn/11" rel="profile">
  <meta http-equiv="X-UA-Compatible" content="IE=edge">
  <meta http-equiv="content-type" content="text/html; charset=utf-8">

  <!-- Enable responsiveness on mobile devices-->
  <meta name="viewport" content="width=device-width, initial-scale=1.0, maximum-scale=1">

  <title>
    
      Lab 9 - Genome Annotation &middot; BIT150
    
  </title>

  <!-- CSS -->
  <link rel="stylesheet" href="/BIT150_website/public/css/poole.css">
  <link rel="stylesheet" href="/BIT150_website/public/css/syntax.css">
  <link rel="stylesheet" href="/BIT150_website/public/css/lanyon.css">
  <link rel="stylesheet" href="https://fonts.googleapis.com/css?family=PT+Serif:400,400italic,700%7CPT+Sans:400">

  <!-- Icons -->
  <link rel="logo" sizes="144x144" href="/BIT150_website/public/logo.png">
  <link rel="shortcut icon" href="/BIT150_website/public/favicon.ico">

  <!-- RSS -->
  <link rel="alternate" type="application/rss+xml" title="RSS" href="/atom.xml">
</head>


  <body class="theme-base-0g">

    <!-- Target for toggling the sidebar `.sidebar-checkbox` is for regular
     styles, `#sidebar-checkbox` for behavior. -->
<input type="checkbox" class="sidebar-checkbox" id="sidebar-checkbox" checked>

<!-- Toggleable sidebar -->
<div class="sidebar" id="sidebar">
  <div class="sidebar-item">
    <h1>
      <a href="/BIT150_website/">
        BIT150
      </a>
    </h1>
    <p class="lead">The UC Davis - CAES course for the development of fundamental skills in Bioinformatics</p>
  </div>

  <nav class="sidebar-nav">
    <a class="sidebar-nav-item" href="/BIT150_website/">Home</a>

    

    
    
      
        
      
    
      
        
      
    
      
        
          <a class="sidebar-nav-item" href="/BIT150_website/about/">About</a>
        
      
    
      
    
      
    
      
        
          <a class="sidebar-nav-item" href="/BIT150_website/course_syllabus/">BIT 150 Syllabus - Fall 2017</a>
        
      
    
      
        
          <a class="sidebar-nav-item" href="/BIT150_website/discussions/">Discussions</a>
        
      
    
      
    
      
        
          <a class="sidebar-nav-item" href="/BIT150_website/labs/">Labs</a>
        
      
    
      
        
          <a class="sidebar-nav-item" href="/BIT150_website/projects/">Projects</a>
        
      
    
      
        
          <a class="sidebar-nav-item" href="/BIT150_website/schedule/">Schedule</a>
        
      
    
      
        
          <a class="sidebar-nav-item" href="/BIT150_website/supplement/">Supplemental Material</a>
        
      
    

<!--
    <a class="sidebar-nav-item" href="/archive/v1.0.0.zip">Download</a>
    <a class="sidebar-nav-item" href="">GitHub project</a>
-->
    <span class="sidebar-nav-item">Currently v1.0.0</span>
  </nav>

  <div class="sidebar-item">
    <p>
      &copy; 2017, DE Runcie & SM Higdon. All rights reserved.
    </p>
  </div>
</div>


    <!-- Wrap is the content to shift when toggling the sidebar. We wrap the
         content to avoid any CSS collisions with our real content. -->
    <div class="wrap">
      <div class="masthead">
        <div class="container">
          <h3 class="masthead-title">
            <a href="/BIT150_website/" title="Home">BIT150</a>
            <small>Applied Bioinformatics</small>
          </h3>
        </div>
      </div>

      <div class="container content">
        <div class="post">
  <h1 class="post-title">Lab 9 - Genome Annotation</h1>
  <span class="post-date">30 Nov 2017</span>
  <div class="lab">
  <h1 id="overview">Overview</h1>

<ul>
  <li><a href="#getting-setup">Getting Setup</a></li>
  <li><a href="#gene-prediction">Gene Prediction</a>
    <ul>
      <li><a href="#prokka">PROKKA</a></li>
    </ul>
  </li>
  <li><a href="#downstream-analysis">Downstream Analysis</a>
    <ul>
      <li><a href="#pan-genome-analysis-with-roary">Pan-genome analysis</a></li>
      <li><a href="#visualizing-in-r">Analysis in R</a></li>
    </ul>
  </li>
  <li><a href="#lab-9-assignment">Lab 9 Assignment</a></li>
</ul>

<h1 id="laboratory-objectives">Laboratory Objectives</h1>

<ul>
  <li>Familiarize concepts and principles related to genome annotation</li>
  <li>Explore strategies for downstream analysis of genome annotations</li>
  <li>Develop shell scripts to automate the annotation of multiple assemblies</li>
  <li>As part of the assignment, responses to all <strong>Thought Questions</strong> will be required</li>
</ul>

<h1 id="getting-setup">Getting Setup</h1>

<p>Last week, we assembled the genome of <em>E. coli</em> O104:H4 strain TY-2482. This bacterium was identified as a virulent human pathogen that lead to a foodbourne disease outbreak in Europe. Today we will continue working with this organism by annotating its genome at the command line using programs on the <strong>farm cluster</strong>.</p>

<p>To begin:</p>

<ol>
  <li>Login to farm at the command line: <code class="highlighter-rouge">ssh user-id@farm.cse.ucdavis.edu</code> and enter your password.</li>
  <li>Once you are logged in, request the following resources for an interactive job to use in lab today:</li>
</ol>

<div class="highlighter-rouge"><pre class="highlight"><code>  srun -p bit150 -t 4:00:00 --mem=8000 -n 1 -c 3 --pty bash -l
</code></pre>
</div>

<h1 id="gene-prediction">Gene Prediction</h1>

<p>There are two major strategies for predicting genes within genome assemblies: reference based sequence alignment and <em>Ab initio</em> (from the beginning) - we will be exploring the latter. While there are pros and cons to both strategies, the <em>Ab initio</em> strategy is useful when a reference genome does not exist for the organism that has been sequenced. By this point, you are aware that genome features exist in several forms, including:</p>

<ul>
  <li>Protein Coding sequences</li>
  <li>Transfer RNAs (tRNAs)</li>
  <li>Ribosomal RNAs (rRNAs)</li>
  <li>Non-coding RNAs (ncRNAs)</li>
</ul>

<p>Each of these DNA feature types can be modeled based on the tetranucleotide patterns and frequencies associated their occurrence in nature. Consequently, there are several Bioinformatics programs that have been developed to <em>annotate</em> the stretches of genomic DNA that encode a given feature type. Today we will learn how to use the program <strong>PROKKA</strong>, which calls upon several other bioinformatics programs in an integrated fashion to comprehensively annotate a prokaryotic genome.</p>

<h2 id="thought-questions">Thought Questions</h2>

<blockquote>
  <ul>
    <li>Describe potential benefits and limitations associated with using sequence alignment based techniques for genome annotation vs. <em>Ab initio</em> strategies.</li>
  </ul>
</blockquote>

<h2 id="prokka">PROKKA</h2>

<ul>
  <li><a href="https://github.com/tseemann/prokka" target="_blank">Prokka: rapid prokaryotic genome annotation</a></li>
  <li><a href="https://academic.oup.com/bioinformatics/article/30/14/2068/2390517" target="_blank">Prokka - iSCB Bioinformatics article</a></li>
</ul>

<p>Prokka is a bioinformatics program that is commonly used to annotate prokaryotic genomes using minimal computational resources. The major requirement of the input files is that they must be in <em>FASTA</em> format.</p>

<h2 id="programs-used-to-predict-genome-features">Programs used to predict genome features</h2>

<p><img src="/BIT150_website/images/prokka-1.png" alt="" /></p>

<h2 id="a-simplified-schema-of-the-prokka-pipeline">A Simplified Schema of the PROKKA pipeline</h2>

<p><img src="/BIT150_website/images/prokka-2.png" alt="" /></p>

<h2 id="command-line-options"><a href="https://github.com/tseemann/prokka#command-line-options" target="_blank">Command Line Options</a></h2>

<p>To view the command line options, we need to load the prokka module:</p>

<div class="highlighter-rouge"><pre class="highlight"><code>module load tbl2asn prokka
</code></pre>
</div>

<p>Now, type the following:</p>

<div class="highlighter-rouge"><pre class="highlight"><code>prokka
</code></pre>
</div>

<p>This should bring up a menu of the command line options that looks very similar to this:</p>

<p><img src="/BIT150_website/images/prokka-3.png" alt="" target="_blank" /></p>

<h2 id="annotating-your-e-coli-genome-assembly">Annotating your <em>E. coli</em> genome assembly</h2>

<p>Today you will annotate the assembled genome of <em>E. coli</em> O104:H4 strain TY-2482 that was generated last week.</p>

<p>First, create a folder in your home directory to work in for today’s lab:</p>

<div class="highlighter-rouge"><pre class="highlight"><code>mkdir Lab_9
</code></pre>
</div>

<p>Similar to last week, <strong>link-in</strong> the data from the data storage directory for the class:</p>

<div class="highlighter-rouge"><pre class="highlight"><code>ln -s /group/BIT150/lab_8/megahit_results/final.contigs.fa ./ecoli_TY2482.contigs.fa
</code></pre>
</div>

<div class="highlighter-rouge"><pre class="highlight"><code>time prokka --outdir prokka_annotation --prefix E.coli_TY-2482 --cpus 3 megahit_results/final.contigs.fa
</code></pre>
</div>
<p><strong><em>Note: Adding the time command before any command will print the compute time when the task is completed.</em></strong></p>

<p>While this is running, let’s discuss the output formats <strong>(It should take about 5-10 minutes for PROKKA to finish)</strong>.</p>

<h2 id="output-formats">Output Formats</h2>

<p><img src="/BIT150_website/images/prokka-4.png" alt="" target="_blank" /></p>

<p>While many of the output formats serve the purpose of submitting data to <strong>NCBI</strong>, some useful formats for further analysis are the <em>FASTA amino acid</em> (<strong>.faa</strong>) and <em>GFF3 - General Feature Format 3</em> (<strong>.gff</strong>). Let’s do something with the GFF3 file we just created!</p>

<h2 id="thought-questions-1">Thought questions</h2>

<blockquote>
  <ul>
    <li>Describe key differences that are likely to exist among the rules used to define the transition state models underpinning the algorithms used by the various programs that PROKKA uses for comprehensive <em>Ab initio</em> gene prediction.</li>
  </ul>
</blockquote>

<h1 id="downstream-analysis">Downstream Analysis</h1>

<p>Because prokaryotes have relatively small genomes compared to the majority of eukaryotes, a common practice in microbiology is to generate a large number of microbial isolates, extract their DNA and sequence their genomes. This has lead to the generation of thousands of microbial genome sequences, all of which may be compared to one another in countless ways. Today we will explore comparative microbial genomics by providing an introduction to the concept of the <strong>bacterial Pan Genome</strong>. The <strong>Pan Genome is the union</strong> of all genes that are identified among a clade of organisms. The clade may be defined at various levels, however it is commonly defined at the species level for Pan Genome analysis. In its most simplistic breakdown, a Pan Genome may then be categorized into a <strong>Core Genome</strong> and <strong>Accessory Genome</strong>. The Core Genome is comprised by all genes that are present within the majority of the taxa being compared at a designated threshold value (intersection of the union of all genes from all taxa in the comparison), while the Accessory Genome (also referred to as the complement of the intersection) consists of the genes that are not common to all taxa being compared.</p>

<p>Today we will use <strong>Roary</strong> to compare the genome features of the virulent <em>E. coli</em> O104:H4 strain TY-2482 to those of a closely related <em>E. coli</em> isolate that is avirulent.</p>

<h2 id="pan-genome-analysis-with-roary">Pan Genome Analysis with Roary</h2>

<ul>
  <li><a href="http://sanger-pathogens.github.io/Roary/">Roary: The Pan Genome</a></li>
  <li><a href="https://academic.oup.com/bioinformatics/article/31/22/3691/240757">Roary - iSCB Bioinformatics Article</a></li>
</ul>

<p>Roary is a fantastic program for comparing the annotation files for multiple prokaryotic genomes because it achieves a core genome alignment in a relatively short time when using minimal compute requirements (see the paper for more info)! Roary is designed to compare high numbers of microbial genomes, but we will only compare two prokaryotic genomes today.</p>

<p>To begin the Pan Genome analysis, let’s make a new sub-directory inside the <code class="highlighter-rouge">Lab_9</code> directory and move into it:</p>

<div class="highlighter-rouge"><pre class="highlight"><code>mkdir roary
cd roary
</code></pre>
</div>

<p>Now, <strong>link-in</strong> the <strong>GFF3</strong> file for the <em>avirulent reference</em> strain of <em>E. coli O104:H4</em> from the <code class="highlighter-rouge">/group/BIT150/Lab_9</code> folder, and the <strong>GFF3</strong> file from your <strong>prokka_annotation</strong> folder to your present working directory (roary):</p>

<div class="highlighter-rouge"><pre class="highlight"><code>ln -s /group/BIT150/Lab_9/E.coli_O104-H4.gff .
ln -s ../prokka_annotation/E.ecoli_TY2482.gff .
</code></pre>
</div>

<p>At this point, you should have <strong>two GFF files</strong> in your present working directory - check by typing <code class="highlighter-rouge">ls</code>.</p>

<p>To begin using the program, we need to load another module:</p>

<div class="highlighter-rouge"><pre class="highlight"><code>module load bio3
</code></pre>
</div>

<p>Next, let’s bring up the roary options menu:</p>

<div class="highlighter-rouge"><pre class="highlight"><code>roary
</code></pre>
</div>

<p>As you can see, roary has many executable program options for customizing your pan-genome analysis. Today we will keep things fairly straight forward. Run the following command to generate the core gene alignment and associated pan genome output files (<strong>Note</strong>:<em>This takes about 20 minutes</em> to finish):</p>

<div class="highlighter-rouge"><pre class="highlight"><code>time roary -i 90 -e --mafft -p 3 *.gff
</code></pre>
</div>

<blockquote>
  <p>While this is running, begin answering the thought questions, or begin setting up your R Studio environment for the next step.</p>
</blockquote>

<p>Once your Pan Genome analysis is completed, check out the statistics:</p>

<div class="highlighter-rouge"><pre class="highlight"><code>less summary_statistics.txt
</code></pre>
</div>
<p><strong>Q to exit less window</strong></p>

<p>The main file we want to investigate for our comparative analysis is in the <code class="highlighter-rouge">gene_presence_absence.csv</code> file.</p>

<p>Have a look:</p>

<div class="highlighter-rouge"><pre class="highlight"><code>less gene_presence_absence.csv
</code></pre>
</div>

<p>As you can see, the data is easy to interpret here. Let’s analyze the data in R and make a plot.</p>

<h2 id="visualizing-in-r">Visualizing in R</h2>

<p>We will use R Studio to analyze the data. <strong>The objective is to create a simple Venn Diagram</strong> that shows the number of unique genes in the virulent strain of <em>E. coli</em>, the number of unique genes in the avirulent strain of <em>E. coli</em>, and the number of shared genes between them (the Core genome between the two).</p>

<h3 id="getting-started">Getting Started</h3>
<p>First, Launch R Studio, then create a new R project (essentially an entire directory dedicated to a single R analysis project):</p>

<blockquote>
  <p><strong>(File -&gt; New Project -&gt; New Directory -&gt; Empty Project -&gt; <em>Name your directory something like R_roary</em>)</strong>.</p>
</blockquote>

<p>Now, we need to get the file from the HPC server onto our local computers. Launch <strong>File Zilla</strong>, connect to farm as you have done in the past, and transfer the <code class="highlighter-rouge">gene_presence_absence.csv</code> file to the folder on your workstation’s hard-drive that corresponds to the <strong>R project directory</strong> you just created. Once this file is transferred, it should be visible in the R Studio window’s file browser:</p>

<p><img src="/BIT150_website/images/lab9-R-1.png" alt="" target="_blank" /></p>

<p>Create a New Rscript (.R), and save the file with a descriptive name - this file is where you will save all of your work:</p>

<blockquote>
  <p><strong>(File -&gt; New File -&gt; R Script)</strong></p>
</blockquote>

<p><strong>Note</strong>: <em>Recall that commands are run from the R script editor by holding the <strong>control</strong> key and pressing <strong>enter</strong> on a given line or highlighted text</em></p>

<p>The R package we will use to make the Venn Diagram is called, <a href="https://rstudio-pubs-static.s3.amazonaws.com/13301_6641d73cfac741a59c0a851feb99e98b.html" target="_blank">VennDiagram</a>. We will also use the <strong>dplyr</strong> package to filter our dataset.</p>

<p>Load the necessary R Package libraries:</p>

<div class="highlighter-rouge"><pre class="highlight"><code>library(dplyr)
install.packages('VennDiagram')
library(VennDiagram)
</code></pre>
</div>

<p>To begin, read in the <em>csv</em> file as a dataframe, treating the first line as a header and filling in all blank cells with an “NA”:</p>

<div class="highlighter-rouge"><pre class="highlight"><code>roary_data &lt;- read.csv("./gene_presence_absence.csv", header = T, na.strings=c("","NA"))
</code></pre>
</div>

<p>Recall that you may view the dataframe using the <code class="highlighter-rouge">View()</code> command or by clicking the table icon of the <code class="highlighter-rouge">R Environment</code> pane of the R Studio session.</p>

<h3 id="filter-the-data">Filter the Data</h3>
<p>The next step is to identify the number of Accessory Genome genes that correspond to either <em>E. coli</em> O104:H4 strain TY-2482, or the reference strain that is avirulent. To do this, the apporach will be to implement a series of filtering steps in order to produce three new dataframes that represent subsets of the entire roary pan genome dataset that we generated.</p>

<ul>
  <li><strong>Subset the Accessory genome:</strong></li>
</ul>

<div class="highlighter-rouge"><pre class="highlight"><code>accessory_genome &lt;- roary_data %&gt;% filter(No..isolates == 1)
</code></pre>
</div>

<ul>
  <li><strong>Subset the TY-2482 Accessory genome:</strong></li>
</ul>

<div class="highlighter-rouge"><pre class="highlight"><code>TY2482_only &lt;- accessory_genome %&gt;% select(Gene, Annotation, E.coli_TY.2842) # This selects data from only the designated 3 columns

TY2482_only &lt;- TY2482_only[complete.cases(TY2482_only), ] # This removes rows with NA values
</code></pre>
</div>

<ul>
  <li><strong>Subset the Accessory genome of the Reference genome:</strong></li>
</ul>

<div class="highlighter-rouge"><pre class="highlight"><code>reference_only &lt;- accessory_genome %&gt;% select(Gene, Annotation, E.coli_ref)

reference_only &lt;- reference_only[complete.cases(reference_only), ]
</code></pre>
</div>

<ul>
  <li><strong>Subset the **Core genome**:</strong></li>
</ul>

<div class="highlighter-rouge"><pre class="highlight"><code>core &lt;- roary_data %&gt;% filter(No..isolates == 2)
</code></pre>
</div>

<h3 id="making-the-venn-diagram">Making the Venn Diagram</h3>

<p>Each row in the subsetted datasets represents an annotated gene. Thus, we can count the row numbers of each genome subset, store the values in R-type objects and call on these R environment objects to make our plot:</p>

<ul>
  <li><strong>Count the number of rows in each Pan genome subset:</strong></li>
</ul>

<div class="highlighter-rouge"><pre class="highlight"><code>core_size &lt;- nrow(core)
TY_2482_accessory_size &lt;- nrow(TY_2482_only)
ref_accessory_size &lt;- nrow(reference_only)
</code></pre>
</div>

<ul>
  <li><strong>Make the diagram:</strong></li>
</ul>

<div class="highlighter-rouge"><pre class="highlight"><code>grid.newpage()
draw.pairwise.venn(area1 = TY2482_accessory_size + core_size, area2 = ref_accessory_size + core_size, cross.area = core_size, category = c("Virulent", "Avirulent"))
</code></pre>
</div>

<p>Here we are visualizing the Pan-genome in a simplistic fashion, but this opens the door to further exploration of each isolate’s contribution to the accessory genome of the comparison.</p>

<h3 id="thought-questions-2">Thought Questions</h3>

<blockquote>
  <ul>
    <li>Describe a strategy you could take to investigate the Accessory genome of these two isolates in an attempt to identify causative genes involved in the virulence outbreak.</li>
  </ul>
</blockquote>

<h2 id="lab-9-assignment">Lab 9 Assignment</h2>

<p>For the assignment, create a markdown document that addresses the following and submit the document to canvas in <strong>html</strong> format:</p>

<h3 id="part-1-thought-questions">Part 1: Thought Questions</h3>

<p>Provide a response to each of the three <strong><em>Thought Questions</em></strong> that are presented at the end of major sections in this lab.</p>

<h3 id="part-2-rast-annotation">Part 2: RAST Annotation</h3>

<p>A very widely used web-based tool for microbial genome annotation is <strong>RAST</strong>. Use this tool to generate an annotation for the assembly you annotated in lab with PROKKA. Compare and contrast the number of genome features identified between the Prokka annotation and the RAST annotation.</p>

<ul>
  <li><strong>Create a RAST account by starting from the homepage: <a href="http://rast.nmpdr.org/">RAST</a></strong>
    <ul>
      <li><em>Do this before you leave lab today</em>
<img src="/BIT150_website/images/rast-1.png" alt="" target="_blank" /></li>
    </ul>
  </li>
  <li><strong>Upload your Genome assembly to RAST for annotation</strong></li>
</ul>

<p><img src="/BIT150_website/images/rast-2.png" alt="" target="_blank" /></p>

<ul>
  <li><strong>Download the annotation in gff format</strong></li>
</ul>

<p>Click the <em>View Details</em> button when the job is finished
<img src="/BIT150_website/images/rast-3.png" alt="" target="_blank" /></p>

<p>Select GFF3 file format and download to your local hard drive.
<img src="/BIT150_website/images/rast-4.png" alt="" target="_blank" /></p>

<ul>
  <li>
    <p><strong>Create a Pan Genome with Roary using the RAST annotation GFF file and the reference genome GFF file you used in lab</strong></p>
  </li>
  <li>
    <p><strong>Compare the summary_statistics of the Pan genome from lab and the RAST-annotated PAn Genome</strong></p>
  </li>
  <li>
    <p><strong>Make a new Venn Diagram using with the RAST annotation data and compare it to the one you made in lab</strong></p>
  </li>
</ul>

<hr />

<p><a href="project-3.md">Project 3</a></p>

</div>

</div>

<div class="related">
  <h2>Related Posts</h2>
  <ul class="related-posts">
    
      <li>
        <h3>
          <a href="/BIT150_website/2017/11/16/lab-08/">
            Lab 8 - Genome Assembly
            <small>16 Nov 2017</small>
          </a>
        </h3>
      </li>
    
      <li>
        <h3>
          <a href="/BIT150_website/2017/11/09/lab-07/">
            Lab 7 - Genome-wide association study
            <small>09 Nov 2017</small>
          </a>
        </h3>
      </li>
    
      <li>
        <h3>
          <a href="/BIT150_website/2017/11/02/lab-06/">
            Lab 6 - Genome Browsers
            <small>02 Nov 2017</small>
          </a>
        </h3>
      </li>
    
  </ul>
</div>

      </div>
    </div>

    <label for="sidebar-checkbox" class="sidebar-toggle"></label>

    <script>
      (function(document) {
        var toggle = document.querySelector('.sidebar-toggle');
        var sidebar = document.querySelector('#sidebar');
        var checkbox = document.querySelector('#sidebar-checkbox');

        document.addEventListener('click', function(e) {
          var target = e.target;

          if(!checkbox.checked ||
             sidebar.contains(target) ||
             (target === checkbox || target === toggle)) return;

          checkbox.checked = false;
        }, false);
      })(document);
    </script>
  </body>
</html>
