<!DOCTYPE html>
<html lang="en-us">

  <head>
  <link href="http://gmpg.org/xfn/11" rel="profile">
  <meta http-equiv="X-UA-Compatible" content="IE=edge">
  <meta http-equiv="content-type" content="text/html; charset=utf-8">

  <!-- Enable responsiveness on mobile devices-->
  <meta name="viewport" content="width=device-width, initial-scale=1.0, maximum-scale=1">

  <title>
    
      Lab 4 - Gene interaction data and intro to R &middot; BIT150
    
  </title>

  <!-- CSS -->
  <link rel="stylesheet" href="/BIT150_website/public/css/poole.css">
  <link rel="stylesheet" href="/BIT150_website/public/css/syntax.css">
  <link rel="stylesheet" href="/BIT150_website/public/css/lanyon.css">
  <link rel="stylesheet" href="https://fonts.googleapis.com/css?family=PT+Serif:400,400italic,700%7CPT+Sans:400">

  <!-- Icons -->
  <link rel="logo" sizes="144x144" href="/BIT150_website/public/logo.png">
  <link rel="shortcut icon" href="/BIT150_website/public/favicon.ico">

  <!-- RSS -->
  <link rel="alternate" type="application/rss+xml" title="RSS" href="/atom.xml">
</head>


  <body class="theme-base-0g">

    <!-- Target for toggling the sidebar `.sidebar-checkbox` is for regular
     styles, `#sidebar-checkbox` for behavior. -->
<input type="checkbox" class="sidebar-checkbox" id="sidebar-checkbox" checked>

<!-- Toggleable sidebar -->
<div class="sidebar" id="sidebar">
  <div class="sidebar-item">
    <h1>
      <a href="/BIT150_website/">
        BIT150
      </a>
    </h1>
    <p class="lead">The UC Davis - CAES course for the development of fundamental skills in Bioinformatics</p>
  </div>

  <nav class="sidebar-nav">
    <a class="sidebar-nav-item" href="/BIT150_website/">Home</a>

    

    
    
      
        
      
    
      
        
      
    
      
        
          <a class="sidebar-nav-item" href="/BIT150_website/about/">About</a>
        
      
    
      
    
      
    
      
        
          <a class="sidebar-nav-item" href="/BIT150_website/course_syllabus/">BIT 150 Syllabus - Fall 2017</a>
        
      
    
      
        
          <a class="sidebar-nav-item" href="/BIT150_website/discussions/">Discussions</a>
        
      
    
      
    
      
        
          <a class="sidebar-nav-item" href="/BIT150_website/labs/">Labs</a>
        
      
    
      
        
          <a class="sidebar-nav-item" href="/BIT150_website/projects/">Projects</a>
        
      
    
      
        
          <a class="sidebar-nav-item" href="/BIT150_website/schedule/">Schedule</a>
        
      
    
      
        
          <a class="sidebar-nav-item" href="/BIT150_website/supplement/">Supplemental Material</a>
        
      
    

<!--
    <a class="sidebar-nav-item" href="/archive/v1.0.0.zip">Download</a>
    <a class="sidebar-nav-item" href="">GitHub project</a>
-->
    <span class="sidebar-nav-item">Currently v1.0.0</span>
  </nav>

  <div class="sidebar-item">
    <p>
      &copy; 2017, DE Runcie & SM Higdon. All rights reserved.
    </p>
  </div>
</div>


    <!-- Wrap is the content to shift when toggling the sidebar. We wrap the
         content to avoid any CSS collisions with our real content. -->
    <div class="wrap">
      <div class="masthead">
        <div class="container">
          <h3 class="masthead-title">
            <a href="/BIT150_website/" title="Home">BIT150</a>
            <small>Applied Bioinformatics</small>
          </h3>
        </div>
      </div>

      <div class="container content">
        <div class="post">
  <h1 class="post-title">Lab 4 - Gene interaction data and intro to R</h1>
  <span class="post-date">19 Oct 2017</span>
  <div class="lab">
  <h2 id="intro">Intro</h2>

<p>In this lab, we will learn to work with <strong>table data</strong>. We’ll use data from online databases, and view,
arrange, and analyze the data using the programing language <code class="highlighter-rouge">R</code>. The data that we’ll focus on
is data on gene interactions, and data on gene annotations. Our goal is to develop techniques to 
test the hypothesis:</p>

<blockquote>
  <p>The genes that cause a disease function in the same networks or pathways</p>
</blockquote>

<h2 id="step-1-gene-interaction-data">Step 1: Gene interaction data</h2>

<p>In class, we discussed the <a href="http://www.ebi.ac.uk/Tools/webservices/psicquic/view">PSICQUIC</a> federated database, a system for querying many autonomous
databases of gene interactions through one unified client. PSICQUIC is very powerful. However, its
output is difficult to process because the results from the different databases don’t all have the same
format.</p>

<p>Instead, in Lab today, we’ll use just one of these autonomous databases called <a href="https://string-db.org/">STRING</a>.
String is a relatively recent database that has many nice features.</p>

<ol>
  <li>
    <p>Start by going to the STRING website and doing a query for genes that interact with BRCA1 in humans.</p>
  </li>
  <li>Take a few minutes to explore the results.
    <blockquote>
      <p>Questions:</p>

      <ul>
        <li>What information does STRING provide about each gene (ie Node)?</li>
        <li>What information does STRING provide about each interaction (ie Edge)?</li>
        <li>Why are 10 genes shown? Find a setting that controls how large the set of genes and interactions is.</li>
      </ul>
    </blockquote>
  </li>
  <li>Click the <strong>Exports</strong> button. You’ll see the interaction data in table format. We want to download this data so that we can process it. Click the <strong>download</strong> button for <code class="highlighter-rouge">... as simple tabular text output</code>. Save the file on your desktop as <code class="highlighter-rouge">BRCA1_String.tsv</code>. Open the file with <code class="highlighter-rouge">Notepad++</code>.
    <blockquote>
      <p>Note:</p>

      <ul>
        <li>In a text editor, a <code class="highlighter-rouge">tsv</code> file looks fairly ordered, but is still hard to read. The file would look pretty much like this if you used <code class="highlighter-rouge">head</code> or <code class="highlighter-rouge">cat</code> to view the file in BASH.</li>
        <li>The main thing to notice here is what separates the <code class="highlighter-rouge">fields</code> (or columns). In this file it is just <code class="highlighter-rouge">whitespace</code>. In a <code class="highlighter-rouge">csv</code> file, there would be a comma (<code class="highlighter-rouge">,</code>) separating each field. In a <code class="highlighter-rouge">tsv</code> file, it’s actually the <code class="highlighter-rouge">TAB</code> character (not a space). This is important because then programs that read this file can tell the difference between a <strong>value</strong> with multiple words (“Gene 3”), and two values in different fields (“Gene” “3”).</li>
      </ul>
    </blockquote>
  </li>
  <li>If you haven’t worked with <code class="highlighter-rouge">R</code> before, you would probably open this file in <code class="highlighter-rouge">Excel</code>. Go ahead an do this. <code class="highlighter-rouge">Excel</code> understands the <code class="highlighter-rouge">tsv</code> format, and correctly separates the data into columns.
    <blockquote>
      <p>Question:</p>

      <ul>
        <li>In Excel, can you sort the interactions by their experimental evidence? What is the strongest-supported interaction in this list?</li>
        <li>In Excel, can you find all interactions that involve BRCA1 itself? How many are there?</li>
      </ul>

      <p>Note:</p>

      <ul>
        <li>Excel is relatively easy to use. However, it can cause lots of problems in Bioinformatics! One common problem is that Excel tries to be <strong>too smart</strong>.</li>
        <li>Try adding a new entry for an interaction between BRCA1 and SEPT2. What happens? What problems would this cause?</li>
      </ul>
    </blockquote>
  </li>
</ol>

<h2 id="step-2-intro-to-r">Step 2. Intro to R.</h2>

<p>Instead, let’s load this same data table into <code class="highlighter-rouge">R</code>.</p>

<p>First, open <code class="highlighter-rouge">RStudio</code>.</p>

<p><code class="highlighter-rouge">Rstudio</code> is like an operating system designed for working in <code class="highlighter-rouge">R</code>. The RStudio program has its own Terminal (called <code class="highlighter-rouge">Console</code>), a File Explorer (called <code class="highlighter-rouge">Files</code>), and a text editor.</p>

<p>To use <code class="highlighter-rouge">R</code>, we enter commands in the <code class="highlighter-rouge">Console</code>. This is very much like entering commands in <code class="highlighter-rouge">Bash</code> - just the language is different.</p>

<p>First, we want to change the directory of our <code class="highlighter-rouge">R</code> session to our computer’s desktop where our file is. In <code class="highlighter-rouge">R</code>, the command is:</p>

<div class="highlighter-rouge"><pre class="highlight"><code>setwd('~/Desktop')
</code></pre>
</div>

<p>This accomplishes the same thing as:</p>
<pre><code class="language-{bash}">cd ~/Desktop
</code></pre>

<p>in Bash.</p>

<p>We can see the files in the current directory by typing:</p>
<div class="highlighter-rouge"><pre class="highlight"><code>list.files()
</code></pre>
</div>

<p>This accomplishes the same thing as:</p>
<pre><code class="language-{bash}">ls
</code></pre>

<p>In Bash, the format of a command was:</p>

<div class="highlighter-rouge"><pre class="highlight"><code>command [arguments] [&gt; output_file]
</code></pre>
</div>

<p>If we don’t give an <code class="highlighter-rouge">output_file</code>, Bash writes the output to the screen. 
But we can <strong>capture</strong> the output in a file to store it and use it later.</p>

<p>In R, we refer to <code class="highlighter-rouge">functions</code> instead of <code class="highlighter-rouge">command</code>s or <code class="highlighter-rouge">program</code>s. The general format is:</p>

<div class="highlighter-rouge"><pre class="highlight"><code>output &lt;- function(argument1,argument2,...)
</code></pre>
</div>

<p>Just like Bash, if we don’t include <code class="highlighter-rouge">output &lt;-</code>, R will print the output of the function to the screen.
But if you do, it will capture the output in the <strong>variable</strong> output. 
The language is different, but the idea is really the same! 
The main difference here is that <code class="highlighter-rouge">output_file</code> in Bash is a file on your hard-drive. It’s permanent. 
But <code class="highlighter-rouge">output</code> in <code class="highlighter-rouge">R</code> is just held in memory. When you close <code class="highlighter-rouge">R</code>, <code class="highlighter-rouge">output</code> disappears! 
This is both good and bad. Good, because memory is MUCH faster than the hard-drive. Bad, because you have 
to save your work in some way more permanent.</p>

<h3 id="scripts">Scripts</h3>
<p>Just like in Bash, we can write scripts in R to record the commands that we would type into the <code class="highlighter-rouge">Console</code>.</p>

<p>RStudio gives us a text editor to do this. But it has a nice feature:</p>

<p>Make a new <code class="highlighter-rouge">R Script</code>. Then type this line:</p>

<div class="highlighter-rouge"><pre class="highlight"><code>x &lt;- 3
</code></pre>
</div>

<p>Now, put your cursor on the line and type <code class="highlighter-rouge">CTRL+Enter</code>. RStudio sends the command to the console and runs it!
No more copy + paste!</p>

<p>You can also run a whole script with the <code class="highlighter-rouge">Source</code> button in RStudio, or with the command:</p>

<div class="highlighter-rouge"><pre class="highlight"><code>source('my_file.R`)
</code></pre>
</div>

<p>R Scripts use comments just like Bash scripts. Just use the <code class="highlighter-rouge">#</code> character at the start of the line.</p>

<h3 id="packages">Packages</h3>

<p>Remember that on Farm, we had to type:</p>

<div class="highlighter-rouge"><pre class="highlight"><code>module load bio
</code></pre>
</div>

<p>to tell Bash to prepare the BLAST programs for us to use?</p>

<p>R is the same, but instead of <code class="highlighter-rouge">modules</code>, R uses <code class="highlighter-rouge">packages</code> or <code class="highlighter-rouge">libraries</code>. For today’s work, we need to load two packages:</p>

<div class="highlighter-rouge"><pre class="highlight"><code>library(dplyr) # for working with tables
library(biomaRt) # for downloading data from BioMart
</code></pre>
</div>

<p>If you get an error message for either library, you’ll have to install these packages:</p>

<div class="highlighter-rouge"><pre class="highlight"><code>install.packages('tidyr')
source("https://bioconductor.org/biocLite.R")
biocLite("biomaRt")
</code></pre>
</div>

<h3 id="loading-data">Loading data</h3>
<p>Since R works with variables instead of files, we have to load data before we can use it.</p>

<div class="highlighter-rouge"><pre class="highlight"><code>brca1_data &lt;- read.delim(file='BRCA1_String.tsv',header=TRUE,stringsAsFactors = FALSE)
</code></pre>
</div>

<p>Note in R Studio how as you type this line, it starts making suggestions for you. Enhanced auto-complete!</p>

<blockquote>
  <p>Question:</p>

  <ul>
    <li>What is the function? What are the arguments?</li>
  </ul>
</blockquote>

<p>Nothing happens when we run this function, because we captured the output in the variable <code class="highlighter-rouge">brca1_data</code>.</p>

<p>How do we view this variable? In Unix, we used <code class="highlighter-rouge">echo</code>, <code class="highlighter-rouge">head</code>, <code class="highlighter-rouge">cat</code>, etc.</p>

<p>In R, we can:</p>

<ol>
  <li>Just type the variable name and hit <code class="highlighter-rouge">Enter</code>. This prints all (or most) to the screen</li>
  <li><code class="highlighter-rouge">View(brca1_data)</code>. RStudio’s spreadsheet viewer.</li>
  <li><code class="highlighter-rouge">head(brca1_data)</code>. The first 6 lines</li>
  <li><code class="highlighter-rouge">glimpse(brca1_data)</code></li>
</ol>

<p>Things to note:</p>

<ol>
  <li><code class="highlighter-rouge">Observations</code> are entries = rows.</li>
  <li><code class="highlighter-rouge">Variables</code> are features = columns.</li>
  <li>Each column has a particular data type. <code class="highlighter-rouge">&lt;chr&gt;</code> is <strong>character</strong>, meaning text. <code class="highlighter-rouge">&lt;dbl&gt;</code> and <code class="highlighter-rouge">&lt;int&gt;</code> are numbers. It’s important to check that this was imported correctly.</li>
  <li>The name of the first column didn’t import well (it starts with <code class="highlighter-rouge">X.</code>?) We can change it like this:</li>
</ol>

<div class="highlighter-rouge"><pre class="highlight"><code>colnames(brca1_data)               # gets all the names of the columns
colnames(brca1_data)[1]            # gets only the first column name
colnames(brca1_data)[1] &lt;- 'node1' # assigns the first column name a new value
colnames(brca1_data)               # check that it changed!
</code></pre>
</div>

<h3 id="accessing-data">Accessing data</h3>

<p>We can access a row (a particular interaction) like this:</p>

<div class="highlighter-rouge"><pre class="highlight"><code>slice(brca1_data,1) # row 1
slice(brca1_data,10) # row 10
</code></pre>
</div>

<p>The arguments are: (1) Name of table variable, (2) row number to select</p>

<p>We can <code class="highlighter-rouge">filter</code> the rows like this:</p>

<div class="highlighter-rouge"><pre class="highlight"><code>filter(brca1_data,node1 == 'BRCA1')
filter(brca1_data,experimentally_determined_interaction &gt; 0.5)
</code></pre>
</div>

<p>The arguments are: (1) Name of table variable, (2) filter criteria.</p>

<p>We can access a column (a feature) like this:</p>

<div class="highlighter-rouge"><pre class="highlight"><code>brca1_data$combined_score # get data from combined_score column
select(brca1_data,combined_score,node1) # select two columns
select(brca1_data,1:3) # select first three columns
</code></pre>
</div>

<blockquote>
  <p>Each of these print the whole result to the screen. How would you print only the first 6 lines? The first 10?</p>

  <p>Hint: You can view all possible arguments to a function by typing <code class="highlighter-rouge">?head</code></p>
</blockquote>

<p>We can sort the data like this:</p>

<div class="highlighter-rouge"><pre class="highlight"><code>arrange(brca1_data,experimentally_determined_interaction)  # increasing values
arrange(brca1_data,desc(experimentally_determined_interaction)) # increasing values
</code></pre>
</div>

<p>We can make a new row like this:</p>

<div class="highlighter-rouge"><pre class="highlight"><code>new_interaction &lt;- data_frame(node1='SEPT2', node2='MARCH1')
bind_rows(brca1_data,new_interaction)
</code></pre>
</div>

<p>It worked, but it filled in <code class="highlighter-rouge">NA</code> or <code class="highlighter-rouge">&lt;NA&gt;</code> for all the other columns because we didn’t provide data for those.</p>

<h3 id="saving-tables">Saving tables</h3>

<p>Once we’ve processed data, we have to save it to a new file, or it will be lost when we close R.</p>

<p>To save a table as a new <code class="highlighter-rouge">.tsv</code> file, do:</p>

<div class="highlighter-rouge"><pre class="highlight"><code>write.table(brca1_data,file = 'processed_BRCA1_interactions.tsv',row.names=FALSE,sep='\t')
</code></pre>
</div>

<blockquote>
  <p>Activity:</p>

  <p>Select only interactions that include BRCA1, have experimentally_determined_interaction &gt; 0.5,
and make a new table including only <code class="highlighter-rouge">node1</code>, <code class="highlighter-rouge">node2</code>, and <code class="highlighter-rouge">experimentally_determined_interaction</code> as columns. Save this as a file.</p>
</blockquote>

<h2 id="data-integration">Data integration</h2>

<p>No that we’ve seen how to do basic data table processing in R, let’s use these skills for some real bioinformatics.</p>

<p>The BRCA1 is known for its role in Breast Cancer.</p>

<p>Remember our original hypothesis:</p>

<blockquote>
  <p>The genes that cause a disease function in the same networks or pathways</p>
</blockquote>

<p>Today, I’d like to try to address a related, but simpler hypothesis:</p>

<blockquote>
  <p>Genes that interact with each other cause similar diseases</p>
</blockquote>

<blockquote>
  <p>Question:</p>

  <p>We now have a data table listing genes that interact. What do we need to test this hypothesis? Where could we get the data?</p>
</blockquote>

<h2 id="querying-biomart-from-r">Querying BioMart from R</h2>

<p>The Ensembl BioMart of Human Genes has an <strong>attribute</strong> imported from the OMIM database describing diseases associated with each gene.</p>

<p>Go to <a href="http://www.ensembl.org/biomart/martview">BioMart</a>, and do a query for the BRCA1 human gene, and get its MIM disease annotations.</p>

<p>We want to do the following:</p>

<ol>
  <li>Get a list of all the genes in our data that interact directly with BRCA1</li>
  <li>Run a BioMart query for all of these genes, extracting all their MIM disease annotations</li>
  <li>Test if any sets of genes cause the same diseases</li>
  <li>Select interactions that may be important for specific diseases</li>
</ol>

<h3 id="list-genes-that-interact-with-brca1">List genes that interact with BRCA1</h3>

<p>We can get this information from <code class="highlighter-rouge">brca1_data</code></p>

<p>Filter interactions where node1 or node2 is BRCA1</p>

<div class="highlighter-rouge"><pre class="highlight"><code>brca1_interactions &lt;- filter(brca1_data,node1=='BRCA1' | node2 == 'BRCA1')
</code></pre>
</div>

<p>Extract the node1 genes and the node2 genes.</p>

<blockquote>
  <p>Note: These are <strong>exchangeable</strong>. There’s no direction of interaction information here</p>
</blockquote>

<div class="highlighter-rouge"><pre class="highlight"><code>node1_genes &lt;- brca1_interactions$node1  # we want the data
node2_genes &lt;- brca1_interactions$node2
</code></pre>
</div>

<p>Combine these into one list</p>

<div class="highlighter-rouge"><pre class="highlight"><code>all_genes &lt;- c(node1_genes,node2_genes)  # the function `c` appends two lists together
</code></pre>
</div>

<p>Find only the unique genes</p>

<div class="highlighter-rouge"><pre class="highlight"><code>unique_genes &lt;- unique(all_genes)
</code></pre>
</div>

<h3 id="set-up-a-biomart-query">Set up a biomart query</h3>

<p>Load the <code class="highlighter-rouge">biomaRt</code> package if you haven’t:</p>

<div class="highlighter-rouge"><pre class="highlight"><code>library(biomaRt)
</code></pre>
</div>

<p>There are several steps to this. They follow very closely to what you do on the website.</p>

<h3 id="find-available-marts-ie-data-warehouses">Find available marts (ie data warehouses)</h3>
<div class="highlighter-rouge"><pre class="highlight"><code>listMarts(host = 'www.ensembl.org')
</code></pre>
</div>

<p>We want the <code class="highlighter-rouge">ensembl Genes 90</code> Mart.</p>

<blockquote>
  <p>Note: There are other sites with other Marts.</p>

  <p>ex. “plants.ensembl.org” for plant biomarts</p>
</blockquote>

<p>Select it:</p>

<div class="highlighter-rouge"><pre class="highlight"><code>ensembl &lt;- useMart('ENSEMBL_MART_ENSEMBL')
ensembl
</code></pre>
</div>

<p>The variable <code class="highlighter-rouge">ensembl</code> holds information about the connection between R and Ensembl</p>

<h3 id="select-a-dataset">Select a dataset</h3>

<p>To view available datasets, do:</p>

<div class="highlighter-rouge"><pre class="highlight"><code>listDatasets(ensembl)
</code></pre>
</div>

<p>Find the one we want (<code class="highlighter-rouge">hsapiens_gene_ensembl</code>), and select it:</p>

<div class="highlighter-rouge"><pre class="highlight"><code>ensembl_human &lt;- useDataset('hsapiens_gene_ensembl',mart = ensembl)
ensembl_human
</code></pre>
</div>

<h3 id="select-filters-and-attributes-for-the-query">Select filters and attributes for the query</h3>

<p>To find available filters, do:</p>

<div class="highlighter-rouge"><pre class="highlight"><code>listFilters(ensembl_human)
</code></pre>
</div>

<p>Our gene names are HUGO names (a standardized human-readable gene name). The filter is: <code class="highlighter-rouge">hgnc_symbol</code></p>

<p>To find available attributes, do:</p>

<div class="highlighter-rouge"><pre class="highlight"><code>listAttributes(ensembl_human)
</code></pre>
</div>

<p>We want our results to include both the HUGO ID and the MIM diseases: <code class="highlighter-rouge">mim_morbid_description</code></p>

<h3 id="run-the-query">Run the query</h3>
<div class="highlighter-rouge"><pre class="highlight"><code>mim_results &lt;- getBM(mart = ensembl_human,
                     attributes = c('hgnc_symbol','mim_morbid_description'), # note how we make a list with the `c()` function
                     filters = c('hgnc_symbol'),
                     values = unique_genes # When we run, R will expand our the variable `unique_genes`, just like BASH did.
                     )
View(mim_results)
</code></pre>
</div>

<blockquote>
  <p>Questions:</p>

  <p>Did all genes get annotations? Were all genes identified in Ensembl?
Does each gene have a single disease annotation?</p>
</blockquote>

<h3 id="process-results">Process results</h3>

<p>Now, we want to address the question: do any genes in this set relate to the same disease?</p>

<p>For this, we’ll use one of the most powerful tools in R for processing data: The <strong>split-apply-combine</strong> approach.</p>

<ol>
  <li>Split the data into different groups by rows</li>
  <li>Apply a function separately to each group</li>
  <li>Combine the results back into a new table</li>
</ol>

<p>Here, we will do the following:</p>

<ol>
  <li>Split the annotation data by disease</li>
  <li>For each disease, collect all genes associated</li>
  <li>Combine into a table, one row for each disease, with all associated genes listed, plus a count of how many genes per disease</li>
</ol>

<h4 id="split-the-annotation-data-by-disease">Split the annotation data by disease</h4>

<p>To split the data, we tell R how we want it split:</p>

<div class="highlighter-rouge"><pre class="highlight"><code>mim_results &lt;- group_by(mim_results,mim_morbid_description)
</code></pre>
</div>

<h4 id="apply-a-function-separately-to-each-group-combine-the-results">Apply a function separately to each group, combine the results</h4>

<p>We want to apply two functions to each group:</p>

<ol>
  <li><code class="highlighter-rouge">n()</code>: count how many rows (ie genes)</li>
  <li><code class="highlighter-rouge">toString()</code>: paste all gene names together into a single string (ie long word or sentence)</li>
</ol>

<p>We use the <code class="highlighter-rouge">summarize</code> function:</p>

<div class="highlighter-rouge"><pre class="highlight"><code>mim_summary &lt;- summarize(mim_results,
                          num_genes = n(),
                          gene_list = toString(hgnc_symbol))
View(mim_summary)
</code></pre>
</div>

<h4 id="sort-the-results-to-answer-the-question">Sort the results to answer the question:</h4>
<p>You can also do this in the SpreadSheet view directly</p>

<div class="highlighter-rouge"><pre class="highlight"><code>arrange(mim_summary,desc(num_genes))
</code></pre>
</div>

<blockquote>
  <p>Questions:</p>

  <p>What diseases are associated with multiple genes in this set?
Are there any <strong>interactions</strong> that where both genes cause the same disease? How would you address this question?</p>
</blockquote>

<h3 id="summary">Summary</h3>
<p>To see how the <strong>split-apply-combine</strong> is working in more detail, think of it like a for loop (yes, we can use them in R too!)</p>

<p>Make an empty table to hold results</p>

<div class="highlighter-rouge"><pre class="highlight"><code>summary_table &lt;- data_frame() 
</code></pre>
</div>

<p>Get a list of the diseases:</p>

<div class="highlighter-rouge"><pre class="highlight"><code>disease_list &lt;- unique(mim_results$mim_morbid_description)
</code></pre>
</div>

<p>Write a for loop. It works just like the one in Bash! But the syntax is a bit different.</p>

<div class="highlighter-rouge"><pre class="highlight"><code>for(disease in disease_list){  

   # select the subset of the rows for the current disease
  disease_results &lt;- filter(mim_results,mim_morbid_description == disease) 
  
   # count the number of rows in the table for the current disease
  n &lt;- nrow(disease_results)    

   # paste together all the gene names
  gene_list &lt;- toString(disease_results$hgnc_symbol)    

	# make a small table to hold this disease's results
  disease_table &lt;- data_frame(disease = disease, n = n, gene_list = gene_list) 
  
   # append this disease's results to the full table
  summary_table &lt;- bind_rows(summary_table,disease_table)    

}
View(summary_table)
</code></pre>
</div>

<h3 id="save-the-results-in-a-tsv-file">Save the results in a <code class="highlighter-rouge">.tsv</code> file:</h3>

<div class="highlighter-rouge"><pre class="highlight"><code>write.table(mim_summary,file = 'BRCA1_interactors_diseases.tsv',row.names=FALSE,sep='\t')
</code></pre>
</div>

<h2 id="summary-1">Summary</h2>

<p>Here is some more background</p>

<h3 id="what-is-r">What is <code class="highlighter-rouge">R</code>?</h3>

<p>R is a programing language. It’s primary purpose is for statistical analyses. 
But it’s also very useful for general data science work, including Bioinformatics. 
R is <strong>Open Source</strong>, meaning that anyone can write programs in <code class="highlighter-rouge">R</code>, and share them for free.
Because of this, there are <strong>packages</strong> in R for almost any Bioinformatics task you will want to do.</p>

<p>For our purposes here, the main advantages of processing data tables in R vs Excel are:</p>

<ul>
  <li>We can write scripts to automate tasks</li>
  <li>Scripts serve as documentation, or a lab notebook of what we did</li>
  <li>R doesn’t do anything that you don’t tell it to do (unlike Excel).</li>
  <li>R can process datafiles that are too large for Excel.</li>
  <li>R can connect directly to online databases</li>
</ul>

<p>These are basically the same reasons why we use Bash and Unix rather than Windows for Bioinformatics!</p>

<h3 id="r-resources">R resources:</h3>

<p><a href="https://www.rstudio.com/wp-content/uploads/2015/02/data-wrangling-cheatsheet.pdf">Data Wrangling Cheat Sheet</a></p>

<p><a href="http://www.datacarpentry.org/R-genomics/01-intro-to-R.html">Intro to R tutorial</a></p>

<h2 id="assignment">Assignment</h2>

<p>Repeat the activity we did here with your gene for Project 2.</p>

<ol>
  <li>Take your gene. Use String, or another database if you’d like to download interacting genes.</li>
  <li>Import the data into R, and use R to get a list of all unique genes.</li>
  <li>Use BioMart to download some attribute of the genes (MIM, GO annotation, Reactome reacting, KEGG pathway), 
and ask if multiple of the interacting genes have similar roles.</li>
  <li>Prepare a brief report in markdown of what you found.</li>
</ol>

<p>Going forward, consider analyses like this for Project 2 itself!</p>


</div>

</div>

<div class="related">
  <h2>Related Posts</h2>
  <ul class="related-posts">
    
      <li>
        <h3>
          <a href="/BIT150_website/2017/11/16/lab-08/">
            Lab 8 - Genome Assembly
            <small>16 Nov 2017</small>
          </a>
        </h3>
      </li>
    
      <li>
        <h3>
          <a href="/BIT150_website/2017/11/09/lab-07/">
            Lab 7 - Genome-wide association study
            <small>09 Nov 2017</small>
          </a>
        </h3>
      </li>
    
      <li>
        <h3>
          <a href="/BIT150_website/2017/11/02/lab-06/">
            Lab 6 - Genome Browsers
            <small>02 Nov 2017</small>
          </a>
        </h3>
      </li>
    
  </ul>
</div>

      </div>
    </div>

    <label for="sidebar-checkbox" class="sidebar-toggle"></label>

    <script>
      (function(document) {
        var toggle = document.querySelector('.sidebar-toggle');
        var sidebar = document.querySelector('#sidebar');
        var checkbox = document.querySelector('#sidebar-checkbox');

        document.addEventListener('click', function(e) {
          var target = e.target;

          if(!checkbox.checked ||
             sidebar.contains(target) ||
             (target === checkbox || target === toggle)) return;

          checkbox.checked = false;
        }, false);
      })(document);
    </script>
  </body>
</html>
