<!DOCTYPE html>
<html lang="en-us">

  <head>
  <link href="http://gmpg.org/xfn/11" rel="profile">
  <meta http-equiv="X-UA-Compatible" content="IE=edge">
  <meta http-equiv="content-type" content="text/html; charset=utf-8">

  <!-- Enable responsiveness on mobile devices-->
  <meta name="viewport" content="width=device-width, initial-scale=1.0, maximum-scale=1">

  <title>
    
      Lab 8: Microbial Genome Assembly &middot; BIT150
    
  </title>

  <!-- CSS -->
  <link rel="stylesheet" href="/BIT150_website/public/css/poole.css">
  <link rel="stylesheet" href="/BIT150_website/public/css/syntax.css">
  <link rel="stylesheet" href="/BIT150_website/public/css/lanyon.css">
  <link rel="stylesheet" href="https://fonts.googleapis.com/css?family=PT+Serif:400,400italic,700%7CPT+Sans:400">

  <!-- Icons -->
  <link rel="logo" sizes="144x144" href="/BIT150_website/public/logo.png">
  <link rel="shortcut icon" href="/BIT150_website/public/favicon.ico">

  <!-- RSS -->
  <link rel="alternate" type="application/rss+xml" title="RSS" href="/atom.xml">
</head>


  <body class="theme-base-0g">

    <!-- Target for toggling the sidebar `.sidebar-checkbox` is for regular
     styles, `#sidebar-checkbox` for behavior. -->
<input type="checkbox" class="sidebar-checkbox" id="sidebar-checkbox" checked>

<!-- Toggleable sidebar -->
<div class="sidebar" id="sidebar">
  <div class="sidebar-item">
    <h1>
      <a href="/BIT150_website/">
        BIT150
      </a>
    </h1>
    <p class="lead">The UC Davis - CAES course for the development of fundamental skills in Bioinformatics</p>
  </div>

  <nav class="sidebar-nav">
    <a class="sidebar-nav-item" href="/BIT150_website/">Home</a>

    

    
    
      
        
      
    
      
        
      
    
      
        
          <a class="sidebar-nav-item" href="/BIT150_website/about/">About</a>
        
      
    
      
    
      
    
      
        
          <a class="sidebar-nav-item" href="/BIT150_website/course_syllabus/">BIT 150 Syllabus - Fall 2017</a>
        
      
    
      
        
          <a class="sidebar-nav-item" href="/BIT150_website/discussions/">Discussions</a>
        
      
    
      
    
      
        
          <a class="sidebar-nav-item" href="/BIT150_website/lab-02/">Lab 2 - HPC and Command Line BLAST</a>
        
      
    
      
        
          <a class="sidebar-nav-item" href="/BIT150_website/lab-03/">Lab 3: R, Bioconductor</a>
        
      
    
      
        
          <a class="sidebar-nav-item" href="/BIT150_website/lab-04/">Lab 4: R, dplyr and ggplot2</a>
        
      
    
      
        
          <a class="sidebar-nav-item" href="/BIT150_website/lab-05/">Lab 5: Cytoscape - Co-expression Networks</a>
        
      
    
      
        
          <a class="sidebar-nav-item" href="/BIT150_website/lab-06/">Lab 6: Genome Browsers and Tracks</a>
        
      
    
      
        
          <a class="sidebar-nav-item" href="/BIT150_website/lab-07/">Lab 7: Analyzing Population Genetics Data</a>
        
      
    
      
        
          <a class="sidebar-nav-item active" href="/BIT150_website/lab-08/">Lab 8: Microbial Genome Assembly</a>
        
      
    
      
        
          <a class="sidebar-nav-item" href="/BIT150_website/lab-09/">Lab 9 - Genome Annotation and Comparisons</a>
        
      
    
      
        
          <a class="sidebar-nav-item" href="/BIT150_website/labs/">Labs</a>
        
      
    
      
        
          <a class="sidebar-nav-item" href="/BIT150_website/project-2/">Project 2: Genotype to Phenotype</a>
        
      
    
      
        
          <a class="sidebar-nav-item" href="/BIT150_website/project-3/">Project 3: Bacterial Outbreak</a>
        
      
    
      
        
          <a class="sidebar-nav-item" href="/BIT150_website/projects/">Projects</a>
        
      
    
      
        
          <a class="sidebar-nav-item" href="/BIT150_website/schedule/">Schedule</a>
        
      
    
      
        
          <a class="sidebar-nav-item" href="/BIT150_website/supplement/">Supplemental Material</a>
        
      
    

<!--
    <a class="sidebar-nav-item" href="/archive/v1.0.0.zip">Download</a>
    <a class="sidebar-nav-item" href="">GitHub project</a>
-->
    <span class="sidebar-nav-item">Currently v1.0.0</span>
  </nav>

  <div class="sidebar-item">
    <p>
      &copy; 2017, DE Runcie & SM Higdon. All rights reserved.
    </p>
  </div>
</div>


    <!-- Wrap is the content to shift when toggling the sidebar. We wrap the
         content to avoid any CSS collisions with our real content. -->
    <div class="wrap">
      <div class="masthead">
        <div class="container">
          <h3 class="masthead-title">
            <a href="/BIT150_website/" title="Home">BIT150</a>
            <small>Applied Bioinformatics</small>
          </h3>
        </div>
      </div>

      <div class="container content">
        <div class="page">
  <h1 class="page-title">Lab 8: Microbial Genome Assembly</h1>
  <h1 id="lab-8-microbial-genome-assembly">Lab 8: Microbial Genome Assembly</h1>

<h2 id="downloading-sequence-data">Downloading Sequence Data</h2>

<p>First, let’s create a set of nested directories to work in for today’s lab. We will put all
files and subfolders associated with this lab into the directory <code class="highlighter-rouge">~/lab_08</code></p>

<div class="highlighter-rouge"><pre class="highlight"><code>mkdir -p ~/lab_08/{data,analysis}
</code></pre>
</div>

<p>Next, let’s download the fastq files containing the raw nucleotide reads for the Escherichia coli O104:H4 str. TY-2482 genome from the <a href="http://www.ebi.ac.uk/ena">ENA</a>:</p>

<div class="highlighter-rouge"><pre class="highlight"><code>cd lab_06/data
curl -O -L ftp://ftp.sra.ebi.ac.uk/vol1/fastq/SRR292/SRR292770/SRR292770_1.fastq.gz
curl -O -L ftp://ftp.sra.ebi.ac.uk/vol1/fastq/SRR292/SRR292770/SRR292770_2.fastq.gz
</code></pre>
</div>

<p>Let’s make sure we downloaded uncorrupted files by using the unix command md5sum:</p>

<div class="highlighter-rouge"><pre class="highlight"><code>md5sum SRR292770_1.fastq.gz SRR292770_2.fastq.gz
</code></pre>
</div>

<p>You should see this:</p>

<div class="highlighter-rouge"><pre class="highlight"><code>89efdc20d85c4a16cdee9b223f7c57c6  SRR292770_1.fastq.gz
83f502a4e75aeaa601fa615afd4a78c9  SRR292770_2.fastq.gz
</code></pre>
</div>

<p>Now if you type:</p>

<div class="highlighter-rouge"><pre class="highlight"><code>ls -l
</code></pre>
</div>

<p>you should see something like:</p>

<div class="highlighter-rouge"><pre class="highlight"><code> total 346936
 -rw-rw-r-- 1 ubuntu ubuntu 169620631 Oct 11 23:37 SRR1976948_1.fastq.gz
 -rw-rw-r-- 1 ubuntu ubuntu 185636992 Oct 11 23:38 SRR1976948_2.fastq.gz
</code></pre>
</div>

<p>One problem with these files is that they are writeable - by default, UNIX
makes things writeable by the file owner. This poses an issue with creating
typos or errors in raw data. Let’s fix that before we go on any further:</p>

<div class="highlighter-rouge"><pre class="highlight"><code>chmod u-w *
</code></pre>
</div>

<p>We’ll talk about what these files are below.</p>

<h3 id="copying-data-into-a-working-directory">Copying data into a working directory</h3>

<p>Recall that above we created two separate directories below the <code class="highlighter-rouge">lab_06</code> directory: <code class="highlighter-rouge">data</code> and <code class="highlighter-rouge">analysis</code>.</p>

<p>It is common to store data files in directories that are separate from the directory in which you store output from analysis with bioinformatic programs.</p>

<p>Rather than typing the absolute or relative path to the data files, let’s create a “virtual copy” of the data in the <strong>analysis</strong> directory
by creating a symbolic link to each file:</p>

<div class="highlighter-rouge"><pre class="highlight"><code>cd ../analysis
ln -s ../data/* .
</code></pre>
</div>

<p>These are FASTQ files – let’s take a look at them:</p>

<div class="highlighter-rouge"><pre class="highlight"><code>less SRR1976948_1.fastq.gz
</code></pre>
</div>

<p><strong>(use the spacebar to scroll down, and type ‘q’ to exit ‘less’)</strong></p>

<blockquote>
  <p><strong>Thought Questions:</strong></p>
  <ul>
    <li>Where does the filename come from?</li>
    <li>Why are there <code class="highlighter-rouge">_1</code> and <code class="highlighter-rouge">_2</code> in the file names?</li>
    <li>What does each of the four lines of a read indicate?</li>
    <li>Links:
      <ul>
        <li><a href="http://en.wikipedia.org/wiki/FASTQ_format">FASTQ Format</a></li>
      </ul>
    </li>
  </ul>
</blockquote>

<h2 id="barcode-trimming-and-read-quality-assessmentcontrol">Barcode Trimming and Read Quality Assessment/Control</h2>

<p>Although raw sequence read files deposited in the SRA database are heavily screened, it is still a good idea to verify that the reads being assembled are free of Illumina adapter sequences, indexing barcode sequences and low quality base calls. We will use two programs to conduct a Quality Assessment and impose Quality Control <strong>(QA/QC)</strong>.</p>

<h3 id="fastqc">FastQC</h3>

<p>We’re going to use <a href="http://www.bioinformatics.babraham.ac.uk/projects/fastqc/">FastQC</a> to
summarize the data. <strong>FastQC</strong> is already installed on our HPC as a module -
let’s activate it and run the program on the two fastq files we downloaded:</p>

<div class="highlighter-rouge"><pre class="highlight"><code>module load fastqc
fastqc SRR292770_1.fastq.gz
fastqc SRR292770_2.fastq.gz
</code></pre>
</div>

<p>Now type ‘ls’:</p>

<div class="highlighter-rouge"><pre class="highlight"><code>ls -d *fastqc.zip*
</code></pre>
</div>

<p>to list the files, and you should see:</p>

<div class="highlighter-rouge"><pre class="highlight"><code>SRR1976948_1_fastqc.zip
SRR1976948_2_fastqc.zip
</code></pre>
</div>

<p>Inside each of the fatqc directories you will find reports from the fastqc. You can download these files using your Jupyter Notebook console, if you like;
or you can look at these copies of them:</p>

<ul>
  <li><code class="highlighter-rouge">SRR1976948_1_fastqc/fastqc_report.html &lt;http://2017-ucsc-metagenomics.readthedocs.io/en/latest/_static/SRR1976948_1_fastqc/fastqc_report.html&gt;</code>__</li>
  <li><code class="highlighter-rouge">SRR1976948_2_fastqc/fastqc_report.html &lt;http://2017-ucsc-metagenomics.readthedocs.io/en/latest/_static/SRR1976948_2_fastqc/fastqc_report.html&gt;</code>__</li>
</ul>

<blockquote>
  <p><strong>Thought Questions:</strong></p>
  <ul>
    <li>What should you pay attention to in the FastQC report?</li>
    <li>Which is “better”, file 1 or file 2? Which factors contribute most to high read quality?</li>
    <li>Links:
      <ul>
        <li><a href="http://www.bioinformatics.babraham.ac.uk/projects/fastqc/">FastQC</a></li>
        <li><a href="http://www.youtube.com/watch?v=bz93ReOv87Y">FastQC tutorial video</a></li>
      </ul>
    </li>
  </ul>
</blockquote>

<p>There are several caveats about FastQC - the main one is that it only
calculates certain statistics (like duplicated sequences) for subsets
of the data (e.g. duplicate sequences are only analyzed for the first</p>

<h3 id="trimmomatic">Trimmomatic</h3>

<p>Now we’re going to do some sequence trimming.  We’ll be using
<a href="http://www.usadellab.org/cms/?page=trimmomatic">Trimmomatic</a>, which
(as with fastqc) is already installed on the cluster as a loadable module.</p>

<p>The first thing we’ll need to do is load the module and download the sequences of the adapters to trim off:</p>

<div class="highlighter-rouge"><pre class="highlight"><code>module load trimmomatic
curl -O -L http://dib-training.ucdavis.edu.s3.amazonaws.com/mRNAseq-semi-2015-03-04/TruSeq2-PE.fa
</code></pre>
</div>

<p>Now, to run Trimmomatic:</p>

<div class="highlighter-rouge"><pre class="highlight"><code> TrimmomaticPE R1.fastq.gz \
               R22.fastq.gz \
               R1.qc.fq.gz s1_se \
               R2.qc.fq.gz s2_se \
               ILLUMINACLIP:TruSeq2-PE.fa:2:40:15 \
               LEADING:2 TRAILING:2 \
               SLIDINGWINDOW:4:2 \
               MINLEN:25
</code></pre>
</div>

<p>You should see output that looks like this:</p>

<div class="highlighter-rouge"><pre class="highlight"><code> ...
 Input Read Pairs: 1000000 Both Surviving: 885734 (88.57%) Forward Only Surviving: 114262 (11.43%) Reverse Only Surviving: 4 (0.00%) Dropped: 0 (0.00%)
 TrimmomaticPE: Completed successfully
</code></pre>
</div>

<blockquote>
  <p><strong>Thought Questions:</strong></p>
  <ul>
    <li>How do you figure out what the parameters mean?</li>
    <li>How do you figure out which parameters to use?</li>
    <li>Which adapters do you use?</li>
    <li>What version of Trimmomatic are we using here? (And FastQC?)</li>
    <li>Do you think parameters are different for RNAseq and genomic data sets?</li>
    <li>What’s with these annoyingly long and complicated filenames?</li>
    <li>Why are we running R1 and R2 together?</li>
  </ul>
</blockquote>

<ul>
  <li>
    <p>For a discussion of optimal trimming strategies, see <a href="http://journal.frontiersin.org/Journal/10.3389/fgene.2014.00013/abstract">MacManes, 2014</a>
– it’s about RNAseq but similar arguments should apply to microbial genome
assembly.</p>
  </li>
  <li>
    <p>Links:</p>
    <ul>
      <li><code class="highlighter-rouge">Trimmomatic &lt;http://www.usadellab.org/cms/?page=trimmomatic&gt;</code>__</li>
    </ul>
  </li>
</ul>

<h3 id="fastqc-again">FastQC Again</h3>

<p>Run FastQC again on the trimmed files::</p>

<p>fastqc R1.qc.fq.gz
   fastqc R2.qc.fq.gz</p>

<p>And now view my copies of these files:</p>

<ul>
  <li><code class="highlighter-rouge">R1.qc_fastqc/fastqc_report.html &lt;http://2016-metagenomics-sio.readthedocs.io/en/work/_static/SRR1976948_1.qc_fastqc/fastqc_report.html&gt;</code>__</li>
  <li><code class="highlighter-rouge">R2.qc_fastqc/fastqc_report.html &lt;http://2016-metagenomics-sio.readthedocs.io/en/work/_static/SRR1976948_2.qc_fastqc/fastqc_report.html&gt;</code>__</li>
</ul>

<p>Let’s take a look at the output files::</p>

<p>less R1.qc.fq.gz</p>

<p>(again, use spacebar to scroll, ‘q’ to exit less).</p>

<blockquote>
  <p>Thought Questions:</p>
  <ul>
    <li>Is the quality trimmed data “better” than before?</li>
    <li>Does it matter that you still have adapters!?</li>
  </ul>
</blockquote>

<h2 id="genome-assembly">Genome Assembly</h2>

<p>There are two main strategies for genome assembly using DNA sequencing reads, <strong><em>de novo</em> assembly</strong> and <strong>reference based mapping</strong>. Among the algorithm types used to carry out <strong><em>de novo</em> assembly</strong>, the two most common strategies are <em>De Bruijn Graph Assembly</em> and <em>Overlap Layout Consensus</em>. For more information on the differences between these two, here are some papers:</p>

<ul>
  <li><a href="https://www.ncbi.nlm.nih.gov/pubmed/22184334">Li et al., 2012</a></li>
  <li><a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC2336801/">Zerbino et al., 2008</a></li>
  <li><a href="http://www.sciencedirect.com/science/article/pii/S0888754310000492">Miller et al., 2010</a></li>
</ul>

<h3 id="megahit">MEGAhit</h3>

<p>To assemble the <em>E. coli</em> genome reads we have been working with, we will use <strong>MEGAhit</strong>, which is an ultra-fast short read assembly program that take a <em>De Bruijn Graph</em> approach.</p>

<ul>
  <li>The paper can be found <a href="https://www.ncbi.nlm.nih.gov/pubmed/25609793">here</a></li>
  <li><a href="https://github.com/voutcn/megahit">MEGAhit source code</a></li>
</ul>

<p>We want to assemble the trimmed reads that were generated after running trimmomatic. In the same directory, run the following commands:</p>

<div class="highlighter-rouge"><pre class="highlight"><code>module load megahit
megahit -1 SRR292770.R1.trim.fq -2 SRR292770.R2.trim.fq -o megahit_2482 -t 4
</code></pre>
</div>

<h2 id="assembly-quality">Assembly Quality</h2>

<p>After a <em>de novo</em> assembly has been generated, a question that may come to mind is…</p>

<blockquote>
  <p>How good is my assembly, and how can I measure the quality?</p>
</blockquote>

<p>While there are many metrics used, two of the most common assembly metrics are <strong>N50</strong> and <strong>L50</strong>.</p>

<p>The <strong>N50</strong> value of an assembly is defined as the length of the smallest contig within a subset of the largest assembled contigs whose sum represents 50 percent of the total number of bases assembled.</p>

<p>The <strong>L50</strong> value is then defined as the the minimum number of contigs whose length sum is equal to the N50 value.</p>

<p>To assess these metrics along with many others, we will use the program below:</p>

<h3 id="quast-quality-assessment-tool-for-genome-assemblies">QUAST: QUality ASsessment Tool for Genome Assemblies</h3>
<ul>
  <li>Click <a href="http://quast.sourceforge.net/">here</a> for more information on QUAST.</li>
  <li>The paper can be found <a href="https://academic.oup.com/bioinformatics/article/29/8/1072/228832/QUAST-quality-assessment-tool-for-genome">here</a>.</li>
  <li><a href="https://github.com/ablab/quast">QUAST source code</a></li>
</ul>

<p>To evaluate the quality of your MEGAhit assembly, run the following commands:</p>

<div class="highlighter-rouge"><pre class="highlight"><code>module load bio
quast.py megahit_2482/final.contigs.fa -o quast_2482
</code></pre>
</div>

<blockquote>
  <p>Notice that we loaded a module called <code class="highlighter-rouge">bio</code> rather than one with a name specific to the program we want to run. Loading this custom module will actually add many commonly used bioinformatics program binary files to your <strong>$PATH</strong> so that you can easily use them in your interactive shell. To check out some of these programs, type the following commands after you have loaded the <code class="highlighter-rouge">bio</code> module:</p>
  <ul>
    <li><code class="highlighter-rouge">pip list</code> to view python modules.</li>
    <li><code class="highlighter-rouge">conda list</code> to view <a href="https://bioconda.github.io/">bioconda</a> modules.</li>
  </ul>
</blockquote>

<p>Running <code class="highlighter-rouge">QUAST</code> on our input file containing the TY-2482 assembly contigs from MEGAhit actually generated many files nested within the new directory, <code class="highlighter-rouge">quast_2482</code>. Let’s have a look at the report in tab-separated-value <strong>(tsv)</strong> format:</p>
<div class="highlighter-rouge"><pre class="highlight"><code>less -S quast_2482/report.txt
</code></pre>
</div>
<p>You should see something very similar to the following:</p>

<div class="highlighter-rouge"><pre class="highlight"><code>All statistics are based on contigs of size &gt;= 500 bp, unless otherwise noted (e.g., "# contigs (&gt;= 0 bp)" and "Tota

Assembly                    final.contigs
# contigs (&gt;= 0 bp)         719
# contigs (&gt;= 1000 bp)      369
# contigs (&gt;= 5000 bp)      211
# contigs (&gt;= 10000 bp)     151
# contigs (&gt;= 25000 bp)     60
# contigs (&gt;= 50000 bp)     19
Total length (&gt;= 0 bp)      5210009
Total length (&gt;= 1000 bp)   5065636
Total length (&gt;= 5000 bp)   4702255
Total length (&gt;= 10000 bp)  4259183
Total length (&gt;= 25000 bp)  2792051
Total length (&gt;= 50000 bp)  1285159
# contigs                   463
Largest contig              143131
Total length                5130667
GC (%)                      50.49
N50                         30790
N75                         14105
L50                         52
L75                         117
# N's per 100 kbp           0.00

</code></pre>
</div>

<blockquote>
  <p><strong>Thought Questions:</strong></p>
  <ul>
    <li>Why do we have so many contigs?</li>
    <li>What would be needed to produce an assembly with a total number of contigs under, say 10?</li>
  </ul>
</blockquote>

<h2 id="putting-it-all-together-sbatch-scripts">Putting it all together: sbatch scripts</h2>

<p>Running through this entire process step by step is great for learning how everything works, but what if you need to assemble many different genomes in a short amount of time? A more efficient way to conduct this entire computational analysis is to write a shell script that can be interpreted by SLURM. Let’s take a look at the example script below:</p>

<h3 id="sbatch-script">sbatch script</h3>

<div class="highlighter-rouge"><pre class="highlight"><code><span class="c">#!/bin/bash -l</span>

<span class="c">#SBATCH -p high</span>
<span class="c">#SBATCH -D /home/smhigdon/BIT150</span>
<span class="c">#SBATCH -o /home/smhigdon/BIT150/slurm-log/MEGA_B150-stdout-%j.txt</span>
<span class="c">#SBATCH -e /home/smhigdon/BIT150/slurm-log/MEGA_B150-stderr-%j.txt</span>
<span class="c">#SBATCH -J MEGA_B150</span>
<span class="c">#SBATCH -N 1</span>
<span class="c">#SBATCH -n 8</span>
<span class="c">#SBATCH --mem=16000</span>
<span class="c">#SBATCH -t 24:00:00</span>
<span class="c">#SBATCH --mail-type=END,FAIL</span>
<span class="c">#SBATCH --mail-user=smhigdon@ucdavis.edu</span>

<span class="c"># Name: megahit_BIT150.sh</span>
<span class="c"># Created by: Shawn Higdon</span>
<span class="c"># creation date: April 25, 2017</span>
<span class="c"># This is a pipeline for the assembly of Illumina PEx150 reads for individual microbial genomes.</span>

<span class="c">## Load the required software module</span>

module load bio

<span class="nv">OUTPUT_FOLDER</span><span class="o">=</span>lab_08
mkdir -p <span class="nv">$OUTPUT_FOLDER</span>

<span class="c">#Begin sample processing loop</span>

<span class="k">for </span>R1 <span class="k">in</span> <span class="nv">$OUTPUT_FOLDER</span>/<span class="k">*</span>_1.fastq.gz
<span class="k">do</span>

<span class="c"># Step One: Identify basename of paired end read files from the same genome library sample</span>

<span class="nb">echo</span> <span class="nv">$R1</span>
<span class="nv">R2</span><span class="o">=</span><span class="sb">`</span><span class="nb">echo</span> <span class="nv">$R1</span> | sed <span class="s1">'s/_1/_2/'</span><span class="sb">`</span>
<span class="nv">bname</span><span class="o">=</span><span class="sb">`</span><span class="nb">echo</span> <span class="nv">$R1</span> | sed <span class="s1">'s/_1.\+//'</span><span class="sb">`</span>
<span class="nv">bname</span><span class="o">=</span><span class="sb">`</span>basename <span class="nv">$bname</span><span class="sb">`</span>
<span class="nb">echo</span> <span class="nv">$R2</span>
<span class="nb">echo</span> <span class="nv">$bname</span>
<span class="nv">samplename</span><span class="o">=</span><span class="k">$(</span><span class="nb">echo</span> <span class="nv">$bname</span> | cut -d_ -f1<span class="k">)</span>
<span class="nb">echo</span> <span class="nv">$samplename</span>

<span class="nv">SAMPLE_FOLDER</span><span class="o">=</span><span class="nv">$OUTPUT_FOLDER</span>/<span class="nv">$samplename</span>
mkdir -p <span class="nv">$SAMPLE_FOLDER</span>

<span class="c"># Step Two: Preprocess the raw paired end read files for each genome library using Trimmomatic</span>

trimmomatic PE -threads 8 <span class="nv">$R1</span> <span class="nv">$R2</span> <span class="nv">$SAMPLE_FOLDER</span>/<span class="nv">$samplename</span>.R1.trim.fq <span class="nv">$SAMPLE_FOLDER</span>/<span class="nv">$samplename</span>.orphan_R1.fq <span class="nv">$SAMPLE_FOLDER</span>/<span class="nv">$samplename</span>.R2.trim.fq <span class="nv">$SAMPLE_FOLDER</span>/<span class="nv">$samplename</span>.orphan_R2.fq ILLUMINACLIP:lab_08/TruSeq-PE-all.fa:2:40:15 LEADING:2 TRAILING:2 SLIDINGWINDOW:4:15 MINLEN:30

<span class="c"># Step Three: Assemble the preprocessed read files for each sample using MEGAHIT</span>

<span class="nv">MEGAHIT_FOLDER</span><span class="o">=</span><span class="nv">$SAMPLE_FOLDER</span>/megahit
megahit -1 <span class="nv">$SAMPLE_FOLDER</span>/<span class="nv">$samplename</span>.R1.trim.fq -2 <span class="nv">$SAMPLE_FOLDER</span>/<span class="nv">$samplename</span>.R2.trim.fq -o <span class="nv">$MEGAHIT_FOLDER</span> -t 8
<span class="c"># Step Four: Use Quast to determine the quality of the MEGAHIT assembly</span>

<span class="nv">QUAST_FOLDER</span><span class="o">=</span><span class="nv">$SAMPLE_FOLDER</span>/quast
quast.py <span class="nv">$MEGAHIT_FOLDER</span>/final.contigs.fa -o <span class="nv">$QUAST_FOLDER</span>

<span class="k">done

</span>hostname
<span class="nb">export </span><span class="nv">OMP_NUM_THREADS</span><span class="o">=</span><span class="nv">$SLURM_NTASKS</span>
module load benchmarks
stream
</code></pre>
</div>

<hr />

<blockquote>
  <p><a href="/BIT150_website/project-3/">Project 3</a></p>
</blockquote>

</div>

      </div>
    </div>

    <label for="sidebar-checkbox" class="sidebar-toggle"></label>

    <script>
      (function(document) {
        var toggle = document.querySelector('.sidebar-toggle');
        var sidebar = document.querySelector('#sidebar');
        var checkbox = document.querySelector('#sidebar-checkbox');

        document.addEventListener('click', function(e) {
          var target = e.target;

          if(!checkbox.checked ||
             sidebar.contains(target) ||
             (target === checkbox || target === toggle)) return;

          checkbox.checked = false;
        }, false);
      })(document);
    </script>
  </body>
</html>
